<!DOCTYPE html><html lang="zh-CN" data-theme="light"><head><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1"><title>01 空间转录组 笔记 | coo1heisenberg's Blog</title><noscript>开启JavaScript才能访问本站哦~</noscript><link rel="icon" href="/img/pwa/favicon.ico"><!-- index.css--><link rel="stylesheet" href="/css/index.css?v=1.10.6"><!-- inject head--><link rel="stylesheet" href="https://cdn2.codesign.qq.com/icons/7pOrz0WXB5ZWJPX/latest/iconfont.css"><!-- aplayer--><link rel="stylesheet" href="//open.lightxi.com/cdnjs/ajax/libs/aplayer/1.10.1/APlayer.min.css"><!-- swiper--><link rel="stylesheet" href="//open.lightxi.com/cdnjs/ajax/libs/Swiper/11.0.5/swiper-bundle.min.css"><!-- fancybox ui--><!-- katex--><link rel="stylesheet" href="//open.lightxi.com/cdnjs/ajax/libs/KaTeX/0.16.9/katex.min.css"><!-- Open Graph--><meta name="description" content="空间转录组 启动服务器端jupyter notebook的流程： 1jupyter notebook --ip&amp;#x3D;0.0.0.0 --port&amp;#x3D;8888 --no-browser 空间转录组的流程分析 聚类 不同的功能区域 对比单细胞转录组分析的聚类：细胞类型 降维 主要使用空间信"><!-- pwa--><meta name="apple-mobile-web-app-capable" content="coo1heisenberg's Blog"><meta name="theme-color" content="var(--efu-main)"><meta name="apple-mobile-web-app-status-bar-style" content="var(--efu-main)"><link rel="bookmark" href="/img/pwa/favicon.ico"><link rel="apple-touch-icon" href="/img/pwa/favicon.ico" sizes="180x180"><script>console.log(
    "%c Program: Hexo %c Theme: Solitude %c Version: v1.10.6",
    "border-radius:5px 0 0 5px;padding: 5px 10px;color:white;background:#ff3842;",
    "padding: 5px 10px;color:white;background:#3e9f50;",
    "padding: 5px 10px;color:white;background:#0084ff;border-radius:0 5px 5px 0",
)
</script><script>(()=>{
        const saveToLocal = {
            set: function setWithExpiry(key, value, ttl) {
                if (ttl === 0)
                    return
                const now = new Date()
                const expiryDay = ttl * 86400000
                const item = {
                    value: value,
                    expiry: now.getTime() + expiryDay
                }
                localStorage.setItem(key, JSON.stringify(item))
            },
            get: function getWithExpiry(key) {
                const itemStr = localStorage.getItem(key)
    
                if (!itemStr) {
                    return undefined
                }
                const item = JSON.parse(itemStr)
                const now = new Date()
    
                if (now.getTime() > item.expiry) {
                    localStorage.removeItem(key)
                    return undefined
                }
                return item.value
            }
        };
        window.utils = {
            saveToLocal: saveToLocal,
            getCSS: (url, id = false) => new Promise((resolve, reject) => {
              const link = document.createElement('link')
              link.rel = 'stylesheet'
              link.href = url
              if (id) link.id = id
              link.onerror = reject
              link.onload = link.onreadystatechange = function() {
                const loadState = this.readyState
                if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
                link.onload = link.onreadystatechange = null
                resolve()
              }
              document.head.appendChild(link)
            }),
            getScript: (url, attr = {}) => new Promise((resolve, reject) => {
              const script = document.createElement('script')
              script.src = url
              script.async = true
              script.onerror = reject
              script.onload = script.onreadystatechange = function() {
                const loadState = this.readyState
                if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
                script.onload = script.onreadystatechange = null
                resolve()
              }
    
              Object.keys(attr).forEach(key => {
                script.setAttribute(key, attr[key])
              })
    
              document.head.appendChild(script)
            }),
            addGlobalFn: (key, fn, name = false, parent = window) => {
                const globalFn = parent.globalFn || {}
                const keyObj = globalFn[key] || {}
        
                if (name && keyObj[name]) return
        
                name = name || Object.keys(keyObj).length
                keyObj[name] = fn
                globalFn[key] = keyObj
                parent.globalFn = globalFn
            },
        }
    })()</script><!-- global head--><script>const GLOBAL_CONFIG = {
    root: '/',
    algolia: undefined,
    localsearch: {"preload":false,"path":"/search.xml"},
    runtime: '2024-05-12 00:00:00',
    lazyload: {
        enable: false,
        error: '/img/error_load.webp'
    },
    copyright: {"limit":50,"author":"作者: Coo1heisenberg’s BLOG","link":"链接: ","source":"来源: coo1heisenberg's Blog","info":"著作权归作者所有。 商业转载请联系作者获得授权，非商业转载请注明出处。"},
    highlight: {
        enable: true,
        limit: 200,
        expand: true,
        copy: true,
        syntax: 'highlight.js'
    },
    randomlink: false,
    lang: {"theme":{"dark":"已切换至深色模式","light":"已切换至浅色模式"},"copy":{"success":"复制成功","error":"复制失败"},"backtop":"返回顶部","time":{"day":"天前","hour":"小时前","just":"刚刚","min":"分钟前","month":"个月前"},"f12":"开发者模式已打开，请遵循GPL协议。","totalk":"无需删除空行，直接输入评论即可","search":{"empty":"找不到你查询的内容：${query}","hit":"找到 ${hits} 条结果，用时 ${time} 毫秒","placeholder":"输入关键词快速查找","count":"共 <b>${count}</b> 条结果。"}},
    aside: {
        sayhello: {
            morning: '一日之计在于晨',
            noon: '吃饱了才有力气干活',
            afternoon: '集中精力，攻克难关',
            night: '不要太劳累了，早睡更健康',
            goodnight: '睡个好觉，保证精力充沛',
        },
        sayhello2: [],
    },
    covercolor: {
        enable: false
    },
    comment: false,
    lightbox: 'null',
    post_ai: false,
    right_menu: false,
};</script><!-- page-config head--><script id="config-diff">var PAGE_CONFIG = {
    is_post: true,
    is_page: false,
    is_home: false,
    page: '',
    toc: true,
    comment: false,
    ai_text: false
}</script><meta name="generator" content="Hexo 7.2.0"></head><body id="body"><!-- universe--><canvas id="universe"></canvas><!-- loading--><!-- console--><div id="console"><div class="close-btn" onclick="sco.hideConsole()"><i class="solitude st-close-fill"></i></div><div class="console-card-group"><div class="console-card-group-right"><div class="console-card tags" onclick="sco.hideConsole()"><div class="card-content"><div class="author-content-item-tips">标签</div><div class="author-content-item-title">寻找感兴趣的领域</div></div><div class="card-tag-cloud"><a href="/tags/Python%E9%AB%98%E7%BA%A7/">Python高级<sup>1</sup></a><a href="/tags/Docker%E5%9F%BA%E7%A1%80%E8%AF%AD%E6%B3%95/">Docker基础语法<sup>1</sup></a><a href="/tags/Nginx%E5%9F%BA%E7%A1%80%E8%AF%AD%E6%B3%95/">Nginx基础语法<sup>1</sup></a><a href="/tags/MongoDB/">MongoDB<sup>1</sup></a><a href="/tags/HTML-CSS/">HTML+CSS<sup>1</sup></a><a href="/tags/Redis/">Redis<sup>2</sup></a><a href="/tags/%E8%AE%BE%E8%AE%A1%E6%A8%A1%E5%BC%8F/">设计模式<sup>1</sup></a><a href="/tags/Java%E9%9B%86%E5%90%88%E6%A1%86%E6%9E%B6/">Java集合框架<sup>1</sup></a><a href="/tags/Linux%E5%9F%BA%E7%A1%80%E8%AF%AD%E6%B3%95/">Linux基础语法<sup>1</sup></a><a href="/tags/Mybatis-Plus/">Mybatis-Plus<sup>1</sup></a><a href="/tags/%E7%A9%BA%E9%97%B4%E8%BD%AC%E5%BD%95%E7%BB%84/">空间转录组<sup>1</sup></a><a href="/tags/Java%E5%B9%B6%E5%8F%91%E7%BC%96%E7%A8%8B/">Java并发编程<sup>1</sup></a><a href="/tags/ElasticSearch/">ElasticSearch<sup>1</sup></a><a href="/tags/Python%E5%9F%BA%E7%A1%80/">Python基础<sup>2</sup></a><a href="/tags/MySQL/">MySQL<sup>1</sup></a><a href="/tags/Project/">Project<sup>2</sup></a><a href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/">单细胞<sup>1</sup></a><a href="/tags/Spring/">Spring<sup>1</sup></a></div></div><div class="console-card history" onclick="sco.hideConsole()"><ul class="card-archive-list"><li class="card-archive-list-item"><a class="card-archive-list-link" href="/archives/2024/11/"><span class="card-archive-list-date">2024/11</span><div class="card-archive-list-count-group"><span class="card-archive-list-count">2</span><span class="card-archive-list-count-unit">篇</span></div></a></li><li class="card-archive-list-item"><a class="card-archive-list-link" href="/archives/2024/09/"><span class="card-archive-list-date">2024/09</span><div class="card-archive-list-count-group"><span class="card-archive-list-count">3</span><span class="card-archive-list-count-unit">篇</span></div></a></li><li class="card-archive-list-item"><a class="card-archive-list-link" href="/archives/2024/07/"><span class="card-archive-list-date">2024/07</span><div class="card-archive-list-count-group"><span class="card-archive-list-count">7</span><span class="card-archive-list-count-unit">篇</span></div></a></li><li class="card-archive-list-item"><a class="card-archive-list-link" href="/archives/2024/06/"><span class="card-archive-list-date">2024/06</span><div class="card-archive-list-count-group"><span class="card-archive-list-count">4</span><span class="card-archive-list-count-unit">篇</span></div></a></li><li class="card-archive-list-item"><a class="card-archive-list-link" href="/archives/2024/05/"><span class="card-archive-list-date">2024/05</span><div class="card-archive-list-count-group"><span class="card-archive-list-count">5</span><span class="card-archive-list-count-unit">篇</span></div></a></li></ul></div></div></div><div class="button-group"><div class="console-btn-item"><span class="darkmode_switchbutton" onclick="sco.switchDarkMode()" title="昼夜切换"><i class="solitude st-moon-clear-fill"></i></span></div><div class="console-btn-item" id="consoleHideAside"><span class="asideSwitch" onclick="sco.switchHideAside()" title="边栏显示控制"><i class="solitude st-side-bar-fill"></i></span></div></div><div class="console-mask" onclick="sco.hideConsole()"></div></div><!-- sidebar--><div id="sidebar" style="zoom: 1;"><div id="menu-mask" style="display: none;"></div><div id="sidebar-menus"><div class="site-data"><div class="data-item is-center"><div class="data-item-link"><a href="/archives/"><div class="headline">文章</div><div class="length-num">21</div></a></div></div><div class="data-item is-center"><div class="data-item-link"><a href="/categories/"><div class="headline">分类</div><div class="length-num">7</div></a></div></div><div class="data-item is-center"><div class="data-item-link"><a href="/tags/"><div class="headline">标签</div><div class="length-num">18</div></a></div></div></div><span class="sidebar-menu-item-title">功能</span><div class="sidebar-menu-item"><span class="darkmode_switchbutton menu-child" onclick="sco.switchDarkMode()"><i class="solitude st-moon-clear-fill"></i><span>显示模式</span></span></div><div class="back-menu-list-groups"><div class="back-menu-list-group"><div class="back-menu-list-title">博客主题</div><div class="back-menu-list"><a class="back-menu-item" target="_blank" rel="noopener" href="https://github.com/valor-x/hexo-theme-solitude" title="Solitude"><img class="nolazyload back-menu-item-icon" src="https://7.isyangs.cn/1/65eb200ee4dea-1.png" alt="Solitude"><span class="back-menu-item-text">Solitude</span></a></div></div></div><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><span>首页</span></a></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><span>文库</span></a><ul class="menus_item_child"><li><a class="site-page child" href="/archives/"><i class="solitude  st-folder-fill"></i><span>文章列表</span></a></li><li><a class="site-page child" href="/categories/"><i class="solitude  st-checkbox-multiple-blank-fill"></i><span>全部分类</span></a></li><li><a class="site-page child" href="/tags/"><i class="solitude  st-price-tag-fill"></i><span>全部标签</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><span>友链</span></a><ul class="menus_item_child"><li><a class="site-page child" href="/links/"><i class="solitude  st-group-fill"></i><span>友情链接</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><span>我的</span></a><ul class="menus_item_child"><li><a class="site-page child" href="/tlink/"><i class="solitude  st-tools-fill"></i><span>工具箱</span></a></li><li><a class="site-page child" href="/music/"><i class="solitude  st-disc-fill"></i><span>音乐馆</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><span>关于</span></a><ul class="menus_item_child"><li><a class="site-page child" href="/about/"><i class="solitude  st-contacts-fill"></i><span>关于本站</span></a></li></ul></div></div><span class="sidebar-menu-item-title">标签</span><div class="card-widget card-tags card-archives card-webinfo card-allinfo"><div class="card-tag-cloud"><a href="/tags/Python%E9%AB%98%E7%BA%A7/">Python高级<sup>1</sup></a><a href="/tags/Docker%E5%9F%BA%E7%A1%80%E8%AF%AD%E6%B3%95/">Docker基础语法<sup>1</sup></a><a href="/tags/Nginx%E5%9F%BA%E7%A1%80%E8%AF%AD%E6%B3%95/">Nginx基础语法<sup>1</sup></a><a href="/tags/MongoDB/">MongoDB<sup>1</sup></a><a href="/tags/HTML-CSS/">HTML+CSS<sup>1</sup></a><a href="/tags/Redis/">Redis<sup>2</sup></a><a href="/tags/%E8%AE%BE%E8%AE%A1%E6%A8%A1%E5%BC%8F/">设计模式<sup>1</sup></a><a href="/tags/Java%E9%9B%86%E5%90%88%E6%A1%86%E6%9E%B6/">Java集合框架<sup>1</sup></a><a href="/tags/Linux%E5%9F%BA%E7%A1%80%E8%AF%AD%E6%B3%95/">Linux基础语法<sup>1</sup></a><a href="/tags/Mybatis-Plus/">Mybatis-Plus<sup>1</sup></a><a href="/tags/%E7%A9%BA%E9%97%B4%E8%BD%AC%E5%BD%95%E7%BB%84/">空间转录组<sup>1</sup></a><a href="/tags/Java%E5%B9%B6%E5%8F%91%E7%BC%96%E7%A8%8B/">Java并发编程<sup>1</sup></a><a href="/tags/ElasticSearch/">ElasticSearch<sup>1</sup></a><a href="/tags/Python%E5%9F%BA%E7%A1%80/">Python基础<sup>2</sup></a><a href="/tags/MySQL/">MySQL<sup>1</sup></a><a href="/tags/Project/">Project<sup>2</sup></a><a href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/">单细胞<sup>1</sup></a><a href="/tags/Spring/">Spring<sup>1</sup></a></div></div></div></div><!-- keyboard--><div class="post" id="body-wrap"><header class="post-bg" id="page-header"><nav class="show" id="nav"><div id="nav-group"><div id="blog_name"><div class="back-home-button" tabindex="-1"><i class="back-home-button-icon solitude st-more-fill"></i><div class="back-menu-list-groups"><div class="back-menu-list-group"><div class="back-menu-list-title">博客主题</div><div class="back-menu-list"><a class="back-menu-item" target="_blank" rel="noopener" href="https://github.com/valor-x/hexo-theme-solitude" title="Solitude"><img class="nolazyload back-menu-item-icon" src="https://7.isyangs.cn/1/65eb200ee4dea-1.png" alt="Solitude"><span class="back-menu-item-text">Solitude</span></a></div></div></div></div><a id="site-name" href="/" title="返回博客主页"><span class="title">Solitude</span></a></div><div id="page-name-mask"><div id="page-name"><a id="page-name-text" onclick="sco.toTop()">01 空间转录组 笔记</a></div></div><div id="menus"><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><span>首页</span></a></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><span>文库</span></a><ul class="menus_item_child"><li><a class="site-page child" href="/archives/"><i class="solitude  st-folder-fill"></i><span>文章列表</span></a></li><li><a class="site-page child" href="/categories/"><i class="solitude  st-checkbox-multiple-blank-fill"></i><span>全部分类</span></a></li><li><a class="site-page child" href="/tags/"><i class="solitude  st-price-tag-fill"></i><span>全部标签</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><span>友链</span></a><ul class="menus_item_child"><li><a class="site-page child" href="/links/"><i class="solitude  st-group-fill"></i><span>友情链接</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><span>我的</span></a><ul class="menus_item_child"><li><a class="site-page child" href="/tlink/"><i class="solitude  st-tools-fill"></i><span>工具箱</span></a></li><li><a class="site-page child" href="/music/"><i class="solitude  st-disc-fill"></i><span>音乐馆</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><span>关于</span></a><ul class="menus_item_child"><li><a class="site-page child" href="/about/"><i class="solitude  st-contacts-fill"></i><span>关于本站</span></a></li></ul></div></div></div><div id="nav-left"></div><div id="nav-right"><div class="nav-button" id="travellings_button"><a class="site-page" href="/" title=""><i class="solitude st-train-line"></i></a></div><div class="nav-button" id="search-button"><a class="site-page social-icon search" href="javascript:void(0);" title="搜索"><i class="solitude st-search-line"></i></a></div><div class="nav-button" id="nav-totop" onclick="sco.toTop()"><a class="totopbtn"><i class="solitude st-arrow-up-line"></i><span id="percent">0</span></a></div><div id="toggle-menu"><a class="site-page"><i class="solitude st-menu-line"></i></a></div></div></div></nav><div class="coverdiv" id="coverdiv"><img class="nolazyload" id="post-cover" src="https://coo1heisenberg-blog.oss-cn-shanghai.aliyuncs.com/bio_gene.png" alt="01 空间转录组 笔记"></div><div id="post-info"><div id="post-firstinfo"><div class="meta-firstline"><a class="post-meta-original" title="该文章为原创文章，注意版权协议">原创</a><span class="post-meta-categories"><a class="post-meta-categories" href="/categories/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/">生信分析</a></span><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/%E7%A9%BA%E9%97%B4%E8%BD%AC%E5%BD%95%E7%BB%84/"><span class="tags-name tags-punctuation">空间转录组</span></a></div></div></div></div><h1 class="post-title">01 空间转录组 笔记</h1><div id="post-meta"><div class="meta-secondline"><span class="post-meta-date" title="发布于 2024-09-26 19:40:27"><i class="post-meta-icon solitude st-calendar-todo-fill"></i><time datetime="2024-09-26T11:40:27.000Z">2024-09-26T11:40:27.000Z</time></span><span class="post-meta-date" title="最后更新于 2024-11-09 14:53:17"><i class="post-meta-icon solitude st-refresh-line"></i><time datetime="2024-11-09T06:53:17.413Z">2024-11-09T06:53:17.413Z</time></span><span class="post-meta-wordcount"><span class="post-meta-separator"></span></span></div></div></div><section class="main-hero-waves-area waves-area"><svg class="waves-svg" xmlns="http://www.w3.org/2000/svg" xlink="http://www.w3.org/1999/xlink" viewBox="0 24 150 28" preserveAspectRatio="none" shape-rendering="auto"><defs><path id="gentle-wave" d="M -160 44 c 30 0 58 -18 88 -18 s 58 18 88 18 s 58 -18 88 -18 s 58 18 88 18 v 44 h -352 Z"></path></defs><g class="parallax"><use href="#gentle-wave" x="48" y="0"></use><use href="#gentle-wave" x="48" y="3"></use><use href="#gentle-wave" x="48" y="5"></use><use href="#gentle-wave" x="48" y="7"></use></g></svg></section></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><h1>空间转录组</h1>
<hr>
<p>启动服务器端jupyter notebook的流程：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">jupyter notebook --ip=0.0.0.0 --port=8888 --no-browser</span><br></pre></td></tr></table></figure>
<h2 id="空间转录组的流程分析"><a class="headerlink" href="#空间转录组的流程分析"></a>空间转录组的流程分析</h2>
<hr>
<ol>
<li><strong>聚类</strong>
<ul>
<li>不同的功能区域
<ul>
<li>对比单细胞转录组分析的聚类：细胞类型</li>
</ul>
</li>
</ul>
</li>
<li><strong>降维</strong>
<ul>
<li>主要使用空间信息进行进行降维，如UMAP、t-SNE，同时考虑细胞之间的空间关系，但一般只有在多个切片联合分析才考虑降维
<ul>
<li>对比单细胞转录组分析的降维：基于基因表达矩阵进行降维，通常使用PCA、t-SNE、UMAP等方法</li>
</ul>
</li>
</ul>
</li>
<li><strong>差异分析</strong>
<ul>
<li>主要发现不同空间区域之间差异表达的基因，如SpatiaIDE
<ul>
<li>对比单细胞转录组分析的差异分析：用于发现不同细胞类型之间的差异表达基因，如Seurat的FindAllMarkers</li>
</ul>
</li>
</ul>
</li>
<li><strong>富集分析</strong>
<ul>
<li>主要用于解释在不同区域或细胞类型中富集的生物学过程，如GO分析、KEGG分析
<ul>
<li>对比单细胞转录组分析的富集分析：用于解释在不同细胞类型中富集的生物学过程，如GO分析、KEGG分析</li>
</ul>
</li>
</ul>
</li>
<li><strong>细胞类型</strong>
<ul>
<li>相互糅合在一起，强调统一体
<ul>
<li>对比单细胞转录组分析的细胞类型：批次分开，强调细胞类型的精确</li>
</ul>
</li>
</ul>
</li>
<li><strong>肿瘤细胞区域</strong>
<ul>
<li>图片可以直接识别
<ul>
<li>对比单细胞转录组分析的肿瘤细胞区域：算法推断</li>
</ul>
</li>
</ul>
</li>
<li><strong>细胞类型分布</strong>
<ul>
<li>能够在组织中确定不同细胞类型的空间分布情况
<ul>
<li>对比单细胞转录组分析的细胞类型分布：一个或几个cluster</li>
</ul>
</li>
</ul>
</li>
<li><strong>共定位</strong>
<ul>
<li>通常指的是组织或者细胞区域内同时表达的基因，这可以通过分析不同区域的基因表达矩阵来确定，共定位基因可能在特定的细胞类型或空间区域中表达，从而解释这些区域的生物学功能或发育过程
<ul>
<li>对比单细胞转录组分析的共定位：通常指的是不同细胞类型之间共同表达的基因，这可以通过分析单细胞数据集中不同细胞簇的基因表达模式来确定，共定位基因可能多个细胞类型中表达，从而揭示这些细胞类型之间的功能联系或分析交互作用</li>
</ul>
</li>
</ul>
</li>
</ol>
<hr>
<h2 id="Python基础"><a class="headerlink" href="#Python基础"></a>Python基础</h2>
<hr>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 获取当前路径</span></span><br><span class="line"><span class="keyword">import</span> os</span><br><span class="line"></span><br><span class="line">cur_path = os.getcwd()</span><br><span class="line"><span class="built_in">print</span>(cur_path)</span><br><span class="line"></span><br><span class="line">os.chdir(<span class="string">&quot;/home/oelab/xxxx&quot;</span>)</span><br><span class="line"><span class="built_in">print</span>(os.path)</span><br></pre></td></tr></table></figure>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">!(python --version)</span><br><span class="line"><span class="comment"># !等价于调用shell命令</span></span><br></pre></td></tr></table></figure>
<hr>
<h2 id="Python进阶"><a class="headerlink" href="#Python进阶"></a>Python进阶</h2>
<hr>
<h3 id="DataFrame"><a class="headerlink" href="#DataFrame"></a>DataFrame</h3>
<p><code>pandas</code> 库中的 <code>DataFrame</code> 是一种二维表结构，可以理解为一个类似于 Excel 表格或数据库表的对象，能够存储不同类型的数据。它的特点是带有行标签（index）和列标签（columns），便于数据的组织、操作和分析</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> panads <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> os</span><br><span class="line"></span><br><span class="line"><span class="comment"># 读取类似的表格文件</span></span><br><span class="line">df = pd.read_csv(<span class="string">&#x27;data.csv&#x27;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 显示最大的列</span></span><br><span class="line"><span class="comment"># display是显示的意思，None是全部列输出</span></span><br><span class="line">pd.set_option(<span class="string">&#x27;display.max_columns&#x27;</span>, <span class="literal">None</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 展示数据框的前8行</span></span><br><span class="line"><span class="built_in">print</span>(df.head(<span class="number">8</span>))</span><br><span class="line"></span><br><span class="line"><span class="comment"># 展示数据框的后8行</span></span><br><span class="line"><span class="built_in">print</span>(df.tail(<span class="number">8</span>))</span><br><span class="line"></span><br><span class="line"><span class="comment"># 查看数值列的统计信息</span></span><br><span class="line">df.describe()</span><br><span class="line"></span><br><span class="line"><span class="comment"># 查看DataFrame的形状（行数、列数）</span></span><br><span class="line">df.shape</span><br><span class="line"></span><br><span class="line"><span class="comment"># 查看数据框所有列的名字</span></span><br><span class="line">df.columns</span><br><span class="line"></span><br><span class="line"><span class="comment"># 查看所有的行名</span></span><br><span class="line"><span class="comment"># 如果没有指定行名，默认从0开始，每次加1</span></span><br><span class="line">df.index</span><br><span class="line"></span><br><span class="line"><span class="comment"># 修改指定的第几个行名</span></span><br><span class="line">a = df.index</span><br><span class="line"><span class="comment"># a是把所有行名提取，它的数据类型是index，接下来转换为list列表</span></span><br><span class="line">index_list = a.tolist()</span><br><span class="line">index_list[<span class="number">3</span>] = <span class="string">&#x27;huge&#x27;</span></span><br><span class="line">df.index = index_list</span><br><span class="line">df.index</span><br><span class="line"></span><br><span class="line"><span class="comment"># 修改指定行名</span></span><br><span class="line">df = df.rename(index = &#123;<span class="string">&#x27;AAACAATGTATCTCCCCA&#x27;</span>: <span class="string">&#x27;huge&#x27;</span>&#125;)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 修改列名</span></span><br><span class="line">df = df.rename(columns=&#123;<span class="string">&#x27;in_tissue&#x27;</span>: <span class="string">&#x27;new_in_tissue&#x27;</span>&#125;, inplace=<span class="literal">True</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 直接用中括号提取某一列</span></span><br><span class="line">df[<span class="string">&#x27;clusters&#x27;</span>]</span><br><span class="line"></span><br><span class="line"><span class="comment"># 提取多列</span></span><br><span class="line">df[[<span class="string">&#x27;clusters&#x27;</span>, <span class="string">&#x27;array_row&#x27;</span>, <span class="string">&#x27;array_col&#x27;</span>]]</span><br><span class="line"></span><br><span class="line"><span class="comment"># 提取一行</span></span><br><span class="line">df.loc[<span class="string">&#x27;AAACAATGTATCTCCCCA-1&#x27;</span>]</span><br><span class="line"></span><br><span class="line"><span class="comment"># 提取一列</span></span><br><span class="line">df.loc[:, <span class="string">&#x27;array_row&#x27;</span>]</span><br><span class="line"></span><br><span class="line"><span class="comment"># 提取指定第一行</span></span><br><span class="line">df.iloc[<span class="number">0</span>]</span><br><span class="line"></span><br><span class="line"><span class="comment"># 提取指定的第一列</span></span><br><span class="line">df.iloc[:, <span class="number">0</span>]</span><br><span class="line"></span><br><span class="line"><span class="comment"># 提取第一行第一列</span></span><br><span class="line">df.iloc[<span class="number">0</span>, <span class="number">0</span>]</span><br><span class="line"></span><br><span class="line"><span class="comment"># 删除行、列</span></span><br><span class="line"><span class="comment"># axis = 1：删除列</span></span><br><span class="line"><span class="comment"># axis = 0：删除行</span></span><br><span class="line">df.drop(<span class="string">&#x27;huge&#x27;</span>, axis = <span class="number">0</span>)</span><br><span class="line">df.drop(<span class="string">&#x27;array_row&#x27;</span>, axis = <span class="number">1</span>)</span><br><span class="line">df.drop([<span class="string">&#x27;array_row&#x27;</span>, <span class="string">&#x27;array_col&#x27;</span>], axis = <span class="number">1</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 按照条件进行过滤</span></span><br><span class="line">df_new = df[df[<span class="string">&#x27;clusters&#x27;</span>] &gt; <span class="number">6</span>]</span><br><span class="line"></span><br><span class="line"><span class="comment"># 数据框的合并</span></span><br><span class="line">df1 = pd.DataFrame(&#123;</span><br><span class="line">    <span class="string">&#x27;k1&#x27;</span>: [<span class="string">&#x27;A&#x27;</span>, <span class="string">&#x27;B&#x27;</span>, <span class="string">&#x27;C&#x27;</span>, <span class="string">&#x27;D&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;k2&#x27;</span>: [<span class="string">&#x27;2&#x27;</span>, <span class="string">&#x27;3&#x27;</span>, <span class="string">&#x27;4&#x27;</span>, <span class="string">&#x27;5&#x27;</span>]</span><br><span class="line">&#125;)</span><br><span class="line"></span><br><span class="line">df2 = pd.DataFrame(&#123;</span><br><span class="line">    <span class="string">&#x27;m1&#x27;</span>: [<span class="string">&#x27;A&#x27;</span>, <span class="string">&#x27;B&#x27;</span>, <span class="string">&#x27;C&#x27;</span>, <span class="string">&#x27;D&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;m2&#x27;</span>: [<span class="string">&#x27;2&#x27;</span>, <span class="string">&#x27;3&#x27;</span>, <span class="string">&#x27;4&#x27;</span>, <span class="string">&#x27;5&#x27;</span>]</span><br><span class="line">&#125;)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 取交集合并两个数据框</span></span><br><span class="line">merge1 = pd.merge(df1, df2, left_on = <span class="string">&#x27;k1&#x27;</span>, right_on = <span class="string">&#x27;m1&#x27;</span>, how = <span class="string">&#x27;inner&#x27;</span>)</span><br><span class="line"><span class="comment"># how 存在 inner left right outer</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># concat：指定轴（行或列）将多个数据框连接在一起，通常用于堆叠数据框，适用于数据框结构相同的情况</span></span><br><span class="line">df_con1 = pd.concat([df1, df2], axis = <span class="number">0</span>)</span><br><span class="line">df_con2 = pd.concat([df1, df2], axis = <span class="number">1</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 按cluster列分组</span></span><br><span class="line"><span class="comment"># 类似于R中的table函数</span></span><br><span class="line">df_group1 = df.groupby(<span class="string">&#x27;clusters&#x27;</span>)</span><br><span class="line"><span class="comment"># 查看每个分组的大小</span></span><br><span class="line">df_group1.size()</span><br><span class="line"></span><br><span class="line"><span class="comment"># for循环</span></span><br><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> [<span class="number">0</span>, <span class="number">2</span>, <span class="number">4</span>, <span class="number">6</span>, <span class="number">8</span>]:</span><br><span class="line">    m = i + <span class="number">1</span></span><br><span class="line">    n = m * <span class="number">2</span></span><br><span class="line">    <span class="built_in">print</span>(n)</span><br><span class="line"> </span><br><span class="line"></span><br><span class="line"><span class="comment"># 该缩进必须要缩进，必须整齐</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 函数</span></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">fun_name</span>(<span class="params">arg1, arg2 ...</span>):</span><br><span class="line">    code1</span><br><span class="line">    code2</span><br></pre></td></tr></table></figure>
<hr>
<h3 id="scverse分析框架-AnnData数据结构"><a class="headerlink" href="#scverse分析框架-AnnData数据结构"></a>scverse分析框架 - AnnData数据结构</h3>
<hr>
<p><strong>等同于R中的S4对象，是数据结构</strong></p>
<p>目前单细胞领域的工具分为三种：</p>
<ol>
<li><code>Bioconductor</code>：R语言实现的生物信息学生态</li>
<li><code>Seurat</code>：R语言实现的单细胞分析生态</li>
<li><code>Scverse</code>：基于Python实现的单细胞分析生态
<ul>
<li>单细胞主要用<code>anndata</code></li>
<li>单细胞多组学主要有<code>mudata</code></li>
<li>空间的多组学用<code>SpatialData</code></li>
</ul>
</li>
</ol>
<p><strong>SCVI-Tools(Single Cell Variational Inference Tools)</strong>：是一个开源的Python库，专为单细胞转录组数据的分析而设计</p>
<p><strong>squidpy</strong>：专门用于处理和分析空间分子数据，基于scanpy和anndata，继承了它们的模块化和可拓展性，并专门处理带有空间坐标的数据提供独特的方法，包括整合组织图像进行深度分析</p>
<p><strong>AnnData结构</strong>：</p>
<p><img src="https://coo1heisenberg-blog.oss-cn-shanghai.aliyuncs.com/image-20241106191942302.png" alt="AnnData数据结构" loading="lazy"></p>
<ol>
<li>核心属性 <strong>X 为一个稀疏矩阵</strong>（<code>scipy.sparse.csr.csr_matrix</code>），用于存储矩阵的值</li>
<li>矩阵的行列分别对应观察者 <strong>obs（即细胞条形码），变量 var （即基因）</strong>，这两个属性分别是一个数据框(<code>panads.core.frame.DataFrame</code>)类型用于存储行列的注释信息，其索引就是对应的 obs_names 和 var_names</li>
</ol>
<table>
<thead>
<tr>
<th></th>
<th>功能</th>
<th>类型</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata.X</code></td>
<td>矩阵</td>
<td>numpy矩阵</td>
</tr>
<tr>
<td><code>adata.obs</code></td>
<td>观测量</td>
<td>pandas Dataframe</td>
</tr>
<tr>
<td><code>adata.var</code></td>
<td>特征量</td>
<td>pandas Dataframe</td>
</tr>
<tr>
<td><code>adata.uns</code></td>
<td>非结构化数据</td>
<td>字典dict</td>
</tr>
</tbody>
</table>
<p><strong>X数据槽</strong>：</p>
<ul>
<li>X slot存储了当前用于分析的表达矩阵，一般使用CSR的稀疏矩阵格式</li>
<li>X 的一个数据结构是numpy array，是没有行名列名消息的，行名和列名消息存储在<code>.obs</code>和<code>.var</code>里</li>
</ul>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">adata.X</span><br><span class="line"></span><br><span class="line"><span class="comment"># 稀疏矩阵转稠密矩阵的方法一</span></span><br><span class="line">adata.X.A</span><br><span class="line"><span class="comment"># 稀疏矩阵转稠密矩阵的方法二</span></span><br><span class="line">df1 = adata.X.todense()</span><br><span class="line"><span class="comment"># 稀疏矩阵转数据框</span></span><br><span class="line">df2 = adata.to_df()</span><br></pre></td></tr></table></figure>
<p><strong>Layers数据槽</strong>：</p>
<ul>
<li>Layers本质上是python的字典，每个layer对应一个表达矩阵的二位数组，比如标准化、对数化等等，这些转换后的数据可以存储在layer中</li>
</ul>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 查看layers下哪些数据</span></span><br><span class="line">adata.layers</span><br><span class="line"></span><br><span class="line"><span class="comment"># 增加layer</span></span><br><span class="line">adata.layers[<span class="string">&quot;raw&quot;</span>] = adata.X</span><br><span class="line"></span><br><span class="line"><span class="comment"># 将raw counts数据复制到新的layer</span></span><br><span class="line">adata.layers[<span class="string">&quot;count_per_million&quot;</span>] = adata.layers[<span class="string">&quot;raw&quot;</span>].copy()</span><br><span class="line">adata.layers[<span class="string">&quot;log_transformed&quot;</span>] = np.log1p(adata.X)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 还原回DataFrames，直接使用.to_df()</span></span><br><span class="line">adata.to_df(layer=<span class="string">&quot;log_transformed&quot;</span>)</span><br></pre></td></tr></table></figure>
<p><strong>使用不同的Layer数据画图</strong>：</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 创建一个包含两个子图的画布，并指定画布大小为(10, 2.5)，ax1和ax2分别代表两个子图</span></span><br><span class="line">fig, (ax1, ax2) = plt.subplots(<span class="number">1</span>, <span class="number">2</span>, figsize=(<span class="number">10</span>, <span class="number">2.5</span>))</span><br><span class="line"></span><br><span class="line"><span class="comment"># 定义感兴趣的基因列表，这些基因将在矩阵图中进行展示</span></span><br><span class="line">genes_of_interest = [<span class="string">&quot;COL1A2&quot;</span>, <span class="string">&quot;SYPL1&quot;</span>]</span><br><span class="line"></span><br><span class="line">sc.pl.matrixplot(</span><br><span class="line">    adata, <span class="comment"># 指定要绘制的数据对象</span></span><br><span class="line">    groupby=<span class="string">&quot;clusters&quot;</span>, <span class="comment"># 指定按照细胞类型（由louvain算法聚类得到的结果）进行分组</span></span><br><span class="line">    var_names=genes_of_interest, <span class="comment"># 指定要在矩阵图中展示的基因</span></span><br><span class="line">    layer=<span class="string">&quot;count_per_million&quot;</span>, <span class="comment"># 指定要绘制的数据层，这里选择了CPM(每百万计数)</span></span><br><span class="line">    ax=ax1, <span class="comment"># 指定当前绘制操作在每一个子图中进行</span></span><br><span class="line">    show=<span class="literal">False</span>, <span class="comment"># 设置为不限制绘制结果，因为将在后面手动调整布局</span></span><br><span class="line">)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 给第一个子图设置标题，说明这个子图中展示的是CPM归一化后的数据</span></span><br><span class="line">ax1.set_title(<span class="string">&quot;CPM normalization&quot;</span>)</span><br><span class="line"></span><br><span class="line">sc.pl.matrixplot(</span><br><span class="line">    adata, <span class="comment"># 指定要绘制的数据对象</span></span><br><span class="line">    groupby=<span class="string">&quot;clusters&quot;</span>, <span class="comment"># 指定按照细胞类型（由louvain算法聚类得到的结果）进行分组</span></span><br><span class="line">    var_names=genes_of_interest, <span class="comment"># 指定要在矩阵图中展示的基因</span></span><br><span class="line">    layer=<span class="string">&quot;raw&quot;</span>, <span class="comment"># 指定要绘制的数据层，这里选择了CPM(每百万计数)</span></span><br><span class="line">    ax=ax2, <span class="comment"># 指定当前绘制操作在每一个子图中进行</span></span><br><span class="line">    show=<span class="literal">False</span>, <span class="comment"># 设置为不限制绘制结果，因为将在后面手动调整布局</span></span><br><span class="line">)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 给第一个子图设置标题，说明这个子图中展示的是CPM归一化后的数据</span></span><br><span class="line">ax2.set_title(<span class="string">&quot;row counts&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 调整子图布局，使得子图之间的间距合适</span></span><br><span class="line">plt.tight_layout()</span><br></pre></td></tr></table></figure>
<p><strong>导出表达矩阵</strong>（就是10X单细胞那三个文件）：</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 导出layers中的矩阵</span></span><br><span class="line">mat1 = adata.to_df(layer=<span class="string">&quot;log_transformed&quot;</span>)</span><br><span class="line">mat1.to_csv(<span class="string">&quot;exprMat_X.csv&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 保存为MTX稀疏矩阵格式</span></span><br><span class="line"><span class="comment"># 创建文件夹</span></span><br><span class="line"><span class="keyword">if</span> <span class="keyword">not</span> os.path.isdir(<span class="string">&quot;sparse_matrix&quot;</span>):</span><br><span class="line">    os.makedirs(<span class="string">&quot;sparse_matrix&quot;</span>)</span><br><span class="line">    </span><br><span class="line"><span class="comment"># 导出稀疏矩阵</span></span><br><span class="line">mmwrite(<span class="string">&quot;sparse_matrix/matrix.mtx&quot;</span>, adata.X)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 另存基因名和barcode列表，通常与mtx文件配套保存</span></span><br><span class="line"><span class="keyword">with</span> <span class="built_in">open</span>(<span class="string">&quot;sparse_matrix/features.tsv&quot;</span>, <span class="string">&#x27;w&#x27;</span>) <span class="keyword">as</span> f:</span><br><span class="line">    f.write(<span class="string">&quot;gene_id\n&quot;</span>)</span><br><span class="line">    <span class="keyword">for</span> gene <span class="keyword">in</span> adata.var_names.values:</span><br><span class="line">        f.write(<span class="string">f&quot;<span class="subst">&#123;gene&#125;</span>\n&quot;</span>)</span><br><span class="line"><span class="keyword">with</span> <span class="built_in">open</span>(<span class="string">&quot;sparse_matrix/barcodes.tsv&quot;</span>, <span class="string">&#x27;w&#x27;</span>) <span class="keyword">as</span> f:</span><br><span class="line">    f.write(<span class="string">&quot;barcodes\n&quot;</span>)</span><br><span class="line">    <span class="keyword">for</span> gene <span class="keyword">in</span> adata.obs_names.values:</span><br><span class="line">        f.write(<span class="string">f&quot;<span class="subst">&#123;barcode&#125;</span>\n&quot;</span>)</span><br></pre></td></tr></table></figure>
<p><strong>基础信息的储存</strong>：</p>
<p><strong>obs就是metadata，var就是基因注释的数据框</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 查看细胞注释的内容</span></span><br><span class="line">adata.obs</span><br><span class="line"></span><br><span class="line"><span class="comment"># 增加新的注释信息</span></span><br><span class="line">adata.obs[<span class="string">&quot;is_low_quality&quot;</span>] = adata.obs[<span class="string">&quot;pct_counts_mt&quot;</span>] &gt; <span class="number">0.03</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 检查更新后的信息</span></span><br><span class="line">adata.obs</span><br><span class="line"></span><br><span class="line"><span class="comment"># 查看细胞名字</span></span><br><span class="line">adata.obs_names</span><br><span class="line"></span><br><span class="line"><span class="comment"># 查看基因</span></span><br><span class="line">adata.var</span><br><span class="line"></span><br><span class="line"><span class="comment"># 查看基因名字</span></span><br><span class="line">adata.var_names</span><br><span class="line"></span><br><span class="line"><span class="comment"># 查看多维注释信息</span></span><br><span class="line">adata.obsm</span><br><span class="line"></span><br><span class="line"><span class="comment"># 整理成数据框输出</span></span><br><span class="line">pd.DataFrame(adata.obsm[<span class="string">&#x27;X_pca&#x27;</span>], index = adata.obs_names, </span><br><span class="line">            columns=[<span class="string">f&#x27;PC_<span class="subst">&#123;i&#125;</span>&#x27;</span> <span class="keyword">for</span> i <span class="keyword">in</span> <span class="built_in">range</span>(<span class="number">1</span>, <span class="number">51</span>)])</span><br><span class="line"></span><br><span class="line"><span class="comment"># 查看所有的细胞配对信息</span></span><br><span class="line">adata.obsp</span><br><span class="line"></span><br><span class="line"><span class="comment"># 查看所有非结构化数据</span></span><br><span class="line">adata.uns_keys()</span><br></pre></td></tr></table></figure>
<p><strong>使用不同降维坐标画图</strong>：</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 创建一个图形，并指定大小为（12， 4）</span></span><br><span class="line">plt.figure(figsize=(<span class="number">12</span>, <span class="number">4</span>))</span><br><span class="line"></span><br><span class="line"><span class="comment"># 创建一个包含1行3列的子图网络，并选择第一个子图作为当前操作对象</span></span><br><span class="line">plt.subplot(<span class="number">1</span>, <span class="number">3</span>, <span class="number">1</span>)</span><br><span class="line"><span class="comment"># PCA</span></span><br><span class="line"></span><br><span class="line">df6 = adata.obs</span><br><span class="line">plt.scatter( <span class="comment"># 绘制散点图</span></span><br><span class="line">	x = adata.obsm[<span class="string">&quot;X_pca&quot;</span>][:, <span class="number">0</span>], <span class="comment"># 指定x轴数据为PCA降维后的第一列数据</span></span><br><span class="line">	y = adata.obsm[<span class="string">&quot;X_pca&quot;</span>][:, <span class="number">1</span>], <span class="comment"># 指定y轴数据为PCA降维后的第一列数据</span></span><br><span class="line">	c = adata.obs[<span class="string">&quot;clusters&quot;</span>] == <span class="string">&quot;3&quot;</span>, <span class="comment"># 根据是否属于B细胞类型，选择颜色，True表示B细胞</span></span><br><span class="line">	s = <span class="number">3</span>, <span class="comment"># 设置散点的大小为3</span></span><br><span class="line">	linewidth = <span class="number">0</span>, <span class="comment"># 设置散点边缘宽度为0，即不显示边缘</span></span><br><span class="line">	cmap = <span class="string">&quot;coolwarm&quot;</span>, <span class="comment"># 选择颜色映射</span></span><br><span class="line">)</span><br><span class="line"></span><br><span class="line">plt.title(<span class="string">&quot;PCA&quot;</span>) <span class="comment"># 设置当前子图的标题为“PCA”</span></span><br><span class="line">plt.axis(<span class="string">&quot;off&quot;</span>) <span class="comment"># 关闭坐标轴</span></span><br><span class="line">plt.gca().set_aspect(<span class="string">&quot;equal&quot;</span>) <span class="comment"># 设置坐标轴的纵横比为1，保证散点图中的点不会被拉伸</span></span><br><span class="line"></span><br><span class="line">sc.tl.tsne(adata)</span><br><span class="line"></span><br><span class="line"><span class="comment"># t-SNE</span></span><br><span class="line">plt.subplot(<span class="number">1</span>, <span class="number">3</span>, <span class="number">2</span>)</span><br><span class="line">plt.scatter( <span class="comment"># 绘制散点图</span></span><br><span class="line">	x = adata.obsm[<span class="string">&quot;X_tsne&quot;</span>][:, <span class="number">0</span>], </span><br><span class="line">	y = adata.obsm[<span class="string">&quot;X_tsne&quot;</span>][:, <span class="number">1</span>],</span><br><span class="line">	c = adata.obs[<span class="string">&quot;clusters&quot;</span>] == <span class="string">&quot;3&quot;</span>,</span><br><span class="line">	s = <span class="number">3</span>,</span><br><span class="line">	linewidth = <span class="number">0</span>,</span><br><span class="line">	cmap = <span class="string">&quot;coolwarm&quot;</span>,</span><br><span class="line">)</span><br><span class="line"></span><br><span class="line">plt.title(<span class="string">&quot;t-SNE&quot;</span>)</span><br><span class="line">plt.axis(<span class="string">&quot;off&quot;</span>)</span><br><span class="line">plt.gca().set_aspect(<span class="string">&quot;equal&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># UMAP</span></span><br><span class="line">plt.subplot(<span class="number">1</span>, <span class="number">3</span>, <span class="number">2</span>)</span><br><span class="line">plt.scatter( <span class="comment"># 绘制散点图</span></span><br><span class="line">	x = adata.obsm[<span class="string">&quot;X_umap&quot;</span>][:, <span class="number">0</span>], </span><br><span class="line">	y = adata.obsm[<span class="string">&quot;X_umap&quot;</span>][:, <span class="number">1</span>],</span><br><span class="line">	c = adata.obs[<span class="string">&quot;clusters&quot;</span>] == <span class="string">&quot;3&quot;</span>,</span><br><span class="line">	s = <span class="number">3</span>,</span><br><span class="line">	linewidth = <span class="number">0</span>,</span><br><span class="line">	cmap = <span class="string">&quot;coolwarm&quot;</span>,</span><br><span class="line">)</span><br><span class="line"></span><br><span class="line">plt.title(<span class="string">&quot;UMAP&quot;</span>)</span><br><span class="line">plt.axis(<span class="string">&quot;off&quot;</span>)</span><br><span class="line">plt.gca().set_aspect(<span class="string">&quot;equal&quot;</span>)</span><br></pre></td></tr></table></figure>
<p><strong>数据合并和筛选</strong></p>
<ol>
<li>
<p>两个对象的合并</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> anndata <span class="keyword">as</span> ad</span><br><span class="line">adatas = [adata1, adata2]</span><br><span class="line">adatas = ad.concatenate(adatas)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>metadata更改</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 构建control组和处理组</span></span><br><span class="line"></span><br><span class="line">adata.obs[<span class="string">&#x27;label&#x27;</span>] = <span class="string">&quot;NA&quot;</span></span><br><span class="line">c = [i <span class="keyword">for</span> i, word <span class="keyword">in</span> <span class="built_in">enumerate</span>(<span class="built_in">list</span>(adata.obs[<span class="string">&#x27;batch1&#x27;</span>])) <span class="keyword">if</span> word.startswith(<span class="string">&#x27;A&#x27;</span>)]</span><br><span class="line">obs[<span class="string">&#x27;label&#x27;</span>][c] = <span class="string">&quot;A&quot;</span></span><br><span class="line">c = [i <span class="keyword">for</span> i, word <span class="keyword">in</span> <span class="built_in">enumerate</span>(<span class="built_in">list</span>(adata.obs[<span class="string">&#x27;batch1&#x27;</span>])) <span class="keyword">if</span> word.startswith(<span class="string">&#x27;B&#x27;</span>)]</span><br><span class="line">obs[<span class="string">&#x27;label&#x27;</span>][c] = <span class="string">&quot;B&quot;</span></span><br><span class="line">c = [i <span class="keyword">for</span> i, word <span class="keyword">in</span> <span class="built_in">enumerate</span>(<span class="built_in">list</span>(adata.obs[<span class="string">&#x27;batch1&#x27;</span>])) <span class="keyword">if</span> word.startswith(<span class="string">&#x27;C&#x27;</span>)]</span><br><span class="line">obs[<span class="string">&#x27;label&#x27;</span>][c] = <span class="string">&quot;C&quot;</span></span><br><span class="line">c = [i <span class="keyword">for</span> i, word <span class="keyword">in</span> <span class="built_in">enumerate</span>(<span class="built_in">list</span>(adata.obs[<span class="string">&#x27;batch1&#x27;</span>])) <span class="keyword">if</span> word.startswith(<span class="string">&#x27;Y&#x27;</span>)]</span><br><span class="line">obs[<span class="string">&#x27;label&#x27;</span>][c] = <span class="string">&quot;Y&quot;</span></span><br></pre></td></tr></table></figure>
</li>
<li>
<p>数据保存和metadata添加</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">os.getcwd()</span><br><span class="line">adata.write(<span class="string">&quot;xxx.h5ad&quot;</span>, compression=<span class="string">&#x27;gzip&#x27;</span>)</span><br><span class="line">adata.obs.to_csv(<span class="string">&quot;scanpy_cell_info.csv&quot;</span>, header=<span class="literal">True</span>, index=<span class="literal">True</span>)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>载入数据</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 加入数据</span></span><br><span class="line">anno = pd.read_csv(<span class="string">&quot;scanpy_cell_info.csv&quot;</span>, index_col = <span class="number">1</span>)</span><br><span class="line"><span class="comment"># 加入样本分组的信息</span></span><br><span class="line">adata.obsp[<span class="string">&#x27;cell_groups&#x27;</span>] = anno[<span class="string">&#x27;label&#x27;</span>]</span><br><span class="line"></span><br><span class="line"><span class="comment"># 甚至可以直接赋值 dataframe</span></span><br><span class="line">adata.obs[<span class="string">&#x27;cell_groups1&#x27;</span>] = <span class="built_in">list</span>(anno[<span class="string">&#x27;clusters&#x27;</span>])</span><br><span class="line">adata.obs = anno</span><br><span class="line">df = adata.obs</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>数据的提取</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># : 表示基因的过滤</span></span><br><span class="line">adata = adata[adata.obs.n_genes_by_counts &lt; <span class="number">2500</span>, :]</span><br><span class="line">adata = adata[adata.obs.pct_counts_mt &lt; <span class="number">5</span>, :]</span><br></pre></td></tr></table></figure>
</li>
</ol>
<hr>
<h2 id="空间转录组基本分析流程"><a class="headerlink" href="#空间转录组基本分析流程"></a>空间转录组基本分析流程</h2>
<hr>
<p><strong>流程</strong>：</p>
<ul>
<li>
<p>导包</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> os</span><br><span class="line"><span class="keyword">import</span> anndata</span><br><span class="line"><span class="keyword">import</span> scanpy <span class="keyword">as</span> sc</span><br><span class="line"><span class="keyword">import</span> matplotlib.pyplot <span class="keyword">as</span> plt</span><br><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"><span class="keyword">import</span> seaborn <span class="keyword">as</span> sns</span><br><span class="line"><span class="keyword">import</span> pooch</span><br><span class="line"><span class="keyword">from</span> scipy.sparse <span class="keyword">import</span> csr_matrix</span><br><span class="line"><span class="keyword">from</span> scipy.io <span class="keyword">import</span> mmwrite</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>查看版本号等</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 输出版本号</span></span><br><span class="line">sc.logging.print_versions()</span><br><span class="line"><span class="comment"># 将图片的背景改为白色</span></span><br><span class="line"><span class="comment"># 设置输出图像格式</span></span><br><span class="line">sc.set_figure_params(facecolor=<span class="string">&quot;white&quot;</span>, figsize=(<span class="number">8</span>, <span class="number">8</span>))</span><br><span class="line"><span class="comment"># verbosity 的取值表示测试结果显示的详细程度，数字越大越详细</span></span><br><span class="line"><span class="comment"># errors = 0</span></span><br><span class="line"><span class="comment"># warnings = 1</span></span><br><span class="line"><span class="comment"># info = 2</span></span><br><span class="line"><span class="comment"># hints = 3</span></span><br><span class="line">sc.settings.verbosity = <span class="number">3</span></span><br></pre></td></tr></table></figure>
</li>
<li>
<p>更改当前的路径</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 获取当前工作路径</span></span><br><span class="line"><span class="built_in">print</span>(os.getcwd())</span><br><span class="line"></span><br><span class="line"><span class="comment"># 更改路径</span></span><br><span class="line">os.chdir(<span class="string">&#x27;/home/oelab/RunData2/STM_AD_analyze/data/GSE220442_counts_and_images/&#x27;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">## 更改完再检查一下有没有更改成功</span></span><br><span class="line"><span class="built_in">print</span>(os.getcwd())</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>读取数据</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">adata = sc.read_visium(<span class="string">&#x27;../GSE220442_counts_and_images/counts_and_images/AD1/&#x27;</span>)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>查看AnnData的数据结构</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">adata</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>标记线粒体基因</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 线粒体百分比计算</span></span><br><span class="line"><span class="comment"># 进行质控</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 将基因名去重</span></span><br><span class="line">adata.var_names_make_unique()</span><br><span class="line"></span><br><span class="line"><span class="comment"># 标记线粒体基因</span></span><br><span class="line"><span class="comment"># 在adata的数据里增加了一行mt</span></span><br><span class="line"><span class="string">&quot;&quot;&quot;</span></span><br><span class="line"><span class="string">用于标记 adata 中的线粒体基因。这里假设线粒体基因的名称以 &quot;MT-&quot; 开头，adata.var[&quot;mt&quot;] 会存储一个布尔值数组，标记每个基因是否为线粒体基因</span></span><br><span class="line"><span class="string">&quot;&quot;&quot;</span></span><br><span class="line">adata.var[<span class="string">&quot;mt&quot;</span>] = adata.var_names.<span class="built_in">str</span>.startswith(<span class="string">&quot;MT-&quot;</span>)</span><br><span class="line"><span class="comment"># pp指对空转数据的预处理</span></span><br><span class="line"><span class="string">&quot;&quot;&quot;</span></span><br><span class="line"><span class="string">sc.pp.calculate_qc_metrics 是 Scanpy 中的一个函数，用于计算质量控制指标。</span></span><br><span class="line"><span class="string">参数 qc_vars=[&quot;mt&quot;] 告诉函数将 adata.var[&quot;mt&quot;] 中标记的线粒体基因作为一个单独的质量控制变量。</span></span><br><span class="line"><span class="string">这样可以计算每个细胞中线粒体基因的表达比例等指标，结果会直接添加到 adata.obs 中</span></span><br><span class="line"><span class="string">&quot;&quot;&quot;</span></span><br><span class="line">sc.pp.calculate_qc_metrics(adata, qc_vars=[<span class="string">&quot;mt&quot;</span>], inplace=<span class="literal">True</span>)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>两种赋值方法</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 第一种赋值方式，浅拷贝</span></span><br><span class="line">adata1=adata</span><br><span class="line"></span><br><span class="line"><span class="comment"># 第二种赋值方式，深拷贝</span></span><br><span class="line">adata1=adata.copy()</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>查看质控图</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## 使用 plt.subplots 创建一个包含 1 行、4 列的图形布局，每个子图的大小为 (15, 4)</span></span><br><span class="line">fig, axs = plt.subplots(<span class="number">1</span>, <span class="number">4</span>, figsize=(<span class="number">15</span>, <span class="number">4</span>))</span><br><span class="line"><span class="comment">## 第一个子图中绘制 adata.obs[&quot;total_counts&quot;] 的直方图，用于显示每个细胞的总计数分布情况</span></span><br><span class="line">sns.histplot(adata.obs[<span class="string">&quot;total_counts&quot;</span>], kde=<span class="literal">False</span>, ax=axs[<span class="number">0</span>])</span><br><span class="line"><span class="comment">## 第二个子图中绘制一个限定了范围的总计数直方图，仅包括总计数少于 10000 的细胞，以便更清晰地观察低计数细胞的分布情况</span></span><br><span class="line">sns.histplot(adata.obs[<span class="string">&quot;total_counts&quot;</span>][adata.obs[<span class="string">&quot;total_counts&quot;</span>] &lt; <span class="number">10000</span>],</span><br><span class="line">    kde=<span class="literal">False</span>, bins=<span class="number">40</span>,ax=axs[<span class="number">1</span>],)</span><br><span class="line"><span class="comment">## 第三个子图中绘制 adata.obs[&quot;n_genes_by_counts&quot;] 的直方图，用于显示每个细胞中检测到的基因数分布</span></span><br><span class="line">sns.histplot(adata.obs[<span class="string">&quot;n_genes_by_counts&quot;</span>], kde=<span class="literal">False</span>, bins=<span class="number">60</span>, ax=axs[<span class="number">2</span>])</span><br><span class="line"><span class="comment">## 第四个子图中绘制基因计数少于 4000 的直方图，方便观察低基因计数的细胞分布</span></span><br><span class="line">sns.histplot(adata.obs[<span class="string">&quot;n_genes_by_counts&quot;</span>][adata.obs[<span class="string">&quot;n_genes_by_counts&quot;</span>] &lt; <span class="number">4000</span>],</span><br><span class="line">    kde=<span class="literal">False</span>, bins=<span class="number">60</span>, ax=axs[<span class="number">3</span>],)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>过滤低质量的细胞</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">sc.pp.filter_cells(adata, min_counts=<span class="number">5000</span>)</span><br><span class="line">sc.pp.filter_cells(adata, max_counts=<span class="number">35000</span>)</span><br><span class="line">adata = adata[adata.obs[<span class="string">&quot;pct_counts_mt&quot;</span>] &lt; <span class="number">20</span>].copy()</span><br><span class="line"><span class="built_in">print</span>(<span class="string">f&quot;#cells after MT filter: <span class="subst">&#123;adata.n_obs&#125;</span>&quot;</span>)</span><br><span class="line">sc.pp.filter_genes(adata, min_cells=<span class="number">10</span>)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>数据归一化</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">sc.pp.normalize_total(adata, inplace=<span class="literal">True</span>)</span><br><span class="line">sc.pp.log1p(adata)</span><br><span class="line">sc.pp.highly_variable_genes(adata, flavor=<span class="string">&quot;seurat&quot;</span>, n_top_genes=<span class="number">2000</span>)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>pca降维，umap可视化，leiden的聚类</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">sc.pp.pca(adata)</span><br><span class="line">sc.pp.neighbors(adata)</span><br><span class="line">sc.tl.umap(adata)</span><br><span class="line"><span class="comment"># 类似于R的findclusters</span></span><br><span class="line">sc.tl.leiden(adata, key_added=<span class="string">&quot;clusters&quot;</span> )</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>展示umap图上基因的表达量</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 过滤之后的数据是什么样的</span></span><br><span class="line">plt.rcParams[<span class="string">&quot;figure.figsize&quot;</span>] = (<span class="number">4</span>, <span class="number">4</span>)</span><br><span class="line"><span class="comment"># 将三张图放在一张图里面</span></span><br><span class="line">sc.pl.umap(adata, color=[<span class="string">&quot;total_counts&quot;</span>, <span class="string">&quot;n_genes_by_counts&quot;</span>, <span class="string">&quot;clusters&quot;</span>], wspace=<span class="number">0.4</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 和上面的结果是一样的</span></span><br><span class="line"><span class="comment"># 分别展示三张图的结果</span></span><br><span class="line">sc.pl.umap(adata, color=[<span class="string">&quot;total_counts&quot;</span> ], wspace=<span class="number">0.4</span>)</span><br><span class="line">sc.pl.umap(adata, color=[ <span class="string">&quot;n_genes_by_counts&quot;</span>], wspace=<span class="number">0.4</span>)</span><br><span class="line">sc.pl.umap(adata, color=[ <span class="string">&quot;clusters&quot;</span>], wspace=<span class="number">0.4</span>)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>可视化空间信息</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 查看空转位置上totalcounts和基因数量的分布</span></span><br><span class="line">plt.rcParams[<span class="string">&quot;figure.figsize&quot;</span>] = (<span class="number">8</span>, <span class="number">8</span>)</span><br><span class="line">sc.pl.spatial(adata, img_key=<span class="string">&quot;hires&quot;</span>, color=[<span class="string">&quot;total_counts&quot;</span>, <span class="string">&quot;n_genes_by_counts&quot;</span>])</span><br><span class="line"></span><br><span class="line"><span class="comment"># 查看每个clusters在空间上是如何分布的</span></span><br><span class="line">sc.pl.spatial(adata, img_key=<span class="string">&quot;hires&quot;</span>, color=<span class="string">&quot;clusters&quot;</span>, size=<span class="number">1.5</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 筛选出1和3两个cluster的分布</span></span><br><span class="line">sc.pl.spatial(adata,</span><br><span class="line">    img_key=<span class="string">&quot;hires&quot;</span>,color=<span class="string">&quot;clusters&quot;</span>,</span><br><span class="line">    groups=[<span class="string">&quot;1&quot;</span>, <span class="string">&quot;3&quot;</span>],crop_coord=[<span class="number">7000</span>, <span class="number">10000</span>, <span class="number">0</span>, <span class="number">6000</span>],</span><br><span class="line">    alpha=<span class="number">0.5</span>,size=<span class="number">1.3</span>,)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>差异分析</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 找出来marker基因</span></span><br><span class="line">sc.tl.rank_genes_groups(adata, <span class="string">&quot;clusters&quot;</span>, method=<span class="string">&quot;t-test&quot;</span>)</span><br><span class="line"><span class="comment"># 基于group2进行展示</span></span><br><span class="line">sc.pl.rank_genes_groups_heatmap(adata, groups=<span class="string">&quot;2&quot;</span>, n_genes=<span class="number">10</span>, groupby=<span class="string">&quot;clusters&quot;</span>)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>个性化展示画图</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 查看cluster的信息</span></span><br><span class="line">sc.pl.spatial(adata, img_key=<span class="string">&quot;hires&quot;</span>, color=[<span class="string">&quot;clusters&quot;</span> ])</span><br><span class="line"></span><br><span class="line"><span class="comment"># 查看两个基因表达量的分布</span></span><br><span class="line">sc.pl.spatial(adata, img_key=<span class="string">&quot;hires&quot;</span>, color=[<span class="string">&quot;COL1A2&quot;</span>, <span class="string">&quot;SYPL1&quot;</span>], alpha=<span class="number">0.7</span>)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>对数据进行保存，方便下一次的读取</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">os.chdir(<span class="string">&quot;/home/oelab/RunData2/STM_AD_analyze/&quot;</span>)</span><br><span class="line">adata.write(<span class="string">&quot;GSE220442.data.anndata.h5ad&quot;</span>)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>保存画图</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">fig.savefig(<span class="string">f&#x27;./figures/Visium-HE_<span class="subst">&#123;reg&#125;</span>.pdf&#x27;</span>)</span><br></pre></td></tr></table></figure>
</li>
</ul>
<hr>
<h2 id="scanpy-单个样本的单细胞分析流程"><a class="headerlink" href="#scanpy-单个样本的单细胞分析流程"></a>scanpy 单个样本的单细胞分析流程</h2>
<hr>
<p><strong>基本的分析流程</strong>：</p>
<ol>
<li>对初始的Adata的预处理</li>
<li>主成分分析</li>
<li>计算 neighborhood graph</li>
<li>对 neighborhood graph 进行 embedding</li>
<li>对 neighborhood graph 进行聚类</li>
<li>找到 marker 基因（簇间比较）</li>
<li>根据 marker 基因注释</li>
</ol>
<p><strong>流程</strong>：</p>
<ol>
<li>
<p><strong>scanpy单细胞数据读取</strong></p>
<ul>
<li>公共数据常用的单细胞数据文件格式：
<ul>
<li>Matrix Market Exchange Format (MEX)：
<ul>
<li><code>.mtx</code> 文件，该文件包含了一个稀疏矩阵，其中每一行代表一个基因，每一列代表一个细胞，而每个素则表示基因在相应细胞中的表达量。通常会有三个文件组成：<code>barcodes.tsv</code>（细胞标识）<code>features.tsv</code>（基因标识）和 <code>matrix.mtx</code>（表达矩阵）</li>
</ul>
</li>
<li>Hierarchical Data Format (HDF5)：
<ul>
<li><code>h5</code> 文件，HDF5 是一种用于存储和组织大规模科学数据的文件格式。在单细胞数据中，可以使用HDF5 存储表达矩阵、基因和细胞的元数据等信息</li>
</ul>
</li>
<li>AnnData：
<ul>
<li><code>.h5ad</code> 文件，是基于 HDF5 的文件格式，由 anndata 库定义。它可以包含单细胞数据的表达矩阵、基因和细胞的元数据、样品信息等</li>
</ul>
</li>
<li>Comma-Separated Values（CSV）或文本文件：
<ul>
<li><code>.csv</code> 或 <code>.txt</code> 文件，包含表达矩阵的文本文件。每一行代表一个基因，每一列代表一个细胞，元素是表达量</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>读取数据</strong>（以mtx为例，但是10X单细胞三个文件得解压）</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 导入scanpy库</span></span><br><span class="line"><span class="keyword">import</span> scanpy <span class="keyword">as</span> sc</span><br><span class="line"></span><br><span class="line"><span class="comment"># 读取tab文件</span></span><br><span class="line">data = sc.read_10X_mtx(<span class="string">&quot;/GSE121121&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 查看数据</span></span><br><span class="line">data</span><br><span class="line"></span><br><span class="line">----------------------------------------------------------------------------------------</span><br><span class="line"><span class="comment"># 读取h5文件</span></span><br><span class="line"><span class="comment"># 导入scanpy库</span></span><br><span class="line"><span class="keyword">import</span> scanpy <span class="keyword">as</span> sc</span><br><span class="line"></span><br><span class="line"><span class="comment"># 读取h5文件</span></span><br><span class="line">data = sc.read_10X_h5(<span class="string">&quot;/xxx.h5&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 查看数据</span></span><br><span class="line">data</span><br><span class="line"></span><br><span class="line"><span class="comment"># 读取h5ad文件</span></span><br><span class="line">adata = sc.read_h5ad(<span class="string">&#x27;/xxx.h5ad&#x27;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 读取csv文件</span></span><br><span class="line">data4 = sc.read_csv(<span class="string">&quot;/xxx.csv&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 读取txt文件</span></span><br><span class="line">data5 = sc.read_text(<span class="string">&quot;/xxx.txt&quot;</span>)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>加载模块</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> os</span><br><span class="line"><span class="keyword">import</span> anndata</span><br><span class="line"><span class="keyword">import</span> scanpy <span class="keyword">as</span> sc</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> matplotlib.pyplot <span class="keyword">as</span> plt</span><br><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"><span class="keyword">import</span> seaborn <span class="keyword">as</span> sns</span><br><span class="line"><span class="keyword">import</span> pooch</span><br><span class="line"><span class="keyword">from</span> scipy.sparse <span class="keyword">import</span> csr_matrix</span><br><span class="line"><span class="keyword">from</span> scipy.io <span class="keyword">import</span> mmwrite</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>初始化设置</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">sc.logging.print_versions()</span><br><span class="line">sc.set_figure_params(facecolor=<span class="string">&quot;white&quot;</span>, figsize=(<span class="number">10</span>, <span class="number">10</span>))</span><br><span class="line">sc.settings.verbosity = <span class="number">3</span></span><br></pre></td></tr></table></figure>
</li>
<li>
<p>读取mtx文件</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">print</span>(os.getcwd())</span><br><span class="line">os.chdir(<span class="string">&quot;/home/oelab/RunData2/STM_AD_analyze/data/&quot;</span>)</span><br><span class="line"><span class="built_in">print</span>(os.getcwd())</span><br><span class="line"></span><br><span class="line"><span class="comment"># 声明h5ad用于存储分析结果</span></span><br><span class="line">result_file = <span class="string">&#x27;sample1_h5ad&#x27;</span></span><br><span class="line"></span><br><span class="line">adata = sc.read_10x_mtx(<span class="string">&quot;../data/AD1/&quot;</span>, <span class="comment"># mtx文件所在的目录</span></span><br><span class="line">                       var_names=<span class="string">&#x27;gene_symbols&#x27;</span>, <span class="comment"># 将 gene 作为 var</span></span><br><span class="line">                       cache=<span class="literal">True</span>)  <span class="comment"># 开启缓存读写</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 消除重复的基因</span></span><br><span class="line">adata.var_names_make_unique()</span><br></pre></td></tr></table></figure>
</li>
<li>
<p><strong>对初始adata的预处理</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 绘制了平均表达量前10的基因箱型图</span></span><br><span class="line">sc.pl.highest_expr_genes(adata, n_top=<span class="number">10</span>)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>质量控制</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 基本的过滤(质控)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># pp指 pre process 预处理</span></span><br><span class="line">sc.pp.filter_cells(adata, min_genes=<span class="number">200</span>)</span><br><span class="line">sc.pp.filter_genes(adata, min_cells=<span class="number">3</span>)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>线粒体百分比的计算</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 线粒体百分比的计算</span></span><br><span class="line">adata.var[<span class="string">&#x27;mt&#x27;</span>] = adata.var_names.<span class="built_in">str</span>.startswith(<span class="string">&#x27;MT-&#x27;</span>)</span><br><span class="line">adata.var[<span class="string">&#x27;mt&#x27;</span>]</span><br><span class="line">sc.pp.calculate_qc_metrics(adata, qc_vars=[<span class="string">&#x27;mt&#x27;</span>], percent_top=<span class="literal">None</span>, log1p=<span class="literal">False</span>, inplace=<span class="literal">True</span>)</span><br><span class="line">adata.obs</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>小提琴图展示</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># total_counts：每个细胞的基因总计数（总表达量）</span></span><br><span class="line"><span class="comment"># pct_counts_mt：每个细胞中，线粒体基因表达量占该细胞所有基因表达量的百分比</span></span><br><span class="line"></span><br><span class="line">sc.pl.violin(adata, [<span class="string">&#x27;n_genes_by_counts&#x27;</span>, <span class="string">&#x27;total_counts&#x27;</span>, <span class="string">&#x27;pct_counts_mt&#x27;</span>], </span><br><span class="line">            jitter=<span class="number">0.4</span>, multi_panel=<span class="literal">True</span>)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>进行过滤</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 进行过滤</span></span><br><span class="line">adata = adata[adata.obs.n_genes_by_counts &gt; <span class="number">500</span>, :]</span><br><span class="line">adata = adata[adata.obs.pct_counts_mt &lt; <span class="number">20</span>, :]</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>进行归一化处理</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 进行归一化处理</span></span><br><span class="line">sc.pp.normalize_total(adata, target_sum=<span class="number">1e4</span>)</span><br><span class="line">sc.pp.log1p(adata)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>高变基因的展示</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 高变基因的展示</span></span><br><span class="line">sc.pp.highly_variable_genes(adata, min_mean=<span class="number">0.0125</span>, max_mean=<span class="number">3</span>, min_disp=<span class="number">0.5</span>)</span><br><span class="line">sc.pl.highly_variable_genes(adata)</span><br><span class="line"></span><br><span class="line">adata.raw = adata</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>提取高变基因的表达矩阵</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 提取高变基因的表达矩阵</span></span><br><span class="line">adata = adata[:, adata.var.highly_variable]</span><br><span class="line"></span><br><span class="line"><span class="comment"># 对两个注释进行回归，消除它俩的影响</span></span><br><span class="line">sc.pp.regress_out(adata, [<span class="string">&#x27;total_counts&#x27;</span>, <span class="string">&#x27;pct_counts_mt&#x27;</span>])</span><br><span class="line"></span><br><span class="line"><span class="comment"># 按零均值单位方差标准化数据，并裁剪超过标准差10的细胞</span></span><br><span class="line">sc.pp.scale(adata, max_value=<span class="number">10</span>)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p><strong>主成分分析（PCA）</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">sc.tl.pca(adata, svd_solver=<span class="string">&#x27;arpack&#x27;</span>)</span><br><span class="line"></span><br><span class="line">sc.pl.pca_variance_ratio(adata, log=<span class="literal">True</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 将结果保存到声明的文件results_file</span></span><br><span class="line">adata.write(result_file)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p><strong>计算 neighborhood graph</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 使用PCA来计算细胞的领域图，为了umap降维，也为了聚类</span></span><br><span class="line">sc.pp.neighbors(adata, n_neighbors=<span class="number">10</span>, n_pcs=<span class="number">40</span>)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p><strong>对 neighborhood graph 进行 embedding</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 进行umap的可视化</span></span><br><span class="line">sc.tl.umap(adata)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 进行umap的展示</span></span><br><span class="line">sc.pl.umap(adata, color=[<span class="string">&#x27;NKG7&#x27;</span>])</span><br><span class="line"></span><br><span class="line"><span class="comment"># 可以不适用原始数据</span></span><br><span class="line">sc.pl.umap(adata, color=[<span class="string">&#x27;NKG7&#x27;</span>], use_raw=<span class="literal">False</span>)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p><strong>对 neighborhood graph 进行聚类</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># scanpy推荐leiden图聚类的方法</span></span><br><span class="line">sc.tl.leiden(adata)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 用umap对leiden聚类结果的可视化</span></span><br><span class="line">sc.pl.umap(adata, color=[<span class="string">&#x27;leiden&#x27;</span>, <span class="string">&#x27;NKG7&#x27;</span>])</span><br></pre></td></tr></table></figure>
</li>
<li>
<p><strong>找到marker基因</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 计算每个簇存在较高程度差异的基因排名</span></span><br><span class="line"><span class="comment"># 在计算差异时，最简单最快的方式是t-检验</span></span><br><span class="line">sc.tl.rank_genes_groups(adata, <span class="string">&#x27;leiden&#x27;</span>, method=<span class="string">&#x27;t-test&#x27;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 展示差异基因</span></span><br><span class="line"><span class="comment"># 每个cluster的top20或者top50marker基因的展示</span></span><br><span class="line">sc.pl.rank_genes_groups(adata, n_genes=<span class="number">25</span>, sharey=<span class="literal">False</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 还可以使用wilcoxon的检测方式</span></span><br><span class="line">sc.tl.rank_genes_groups(adata, <span class="string">&#x27;leiden&#x27;</span>, method=<span class="string">&#x27;wilcoxon&#x27;</span>)</span><br><span class="line">sc.pl.rank_genes_groups(adata, n_genes=<span class="number">25</span>, sharey=<span class="literal">False</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 差异基因以表格形式查看</span></span><br><span class="line">pd.DataFrame(adata.uns[<span class="string">&#x27;rank_genes_groups&#x27;</span>][<span class="string">&#x27;names&#x27;</span>]).head(<span class="number">5</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 显示差异基因的名称和差异度</span></span><br><span class="line">sc.tl.rank_genes_groups(adata, <span class="string">&#x27;leiden&#x27;</span>, method=<span class="string">&#x27;wilcoxon&#x27;</span>)</span><br><span class="line">sc.pl.rank_genes_groups(adata, n_genes=<span class="number">25</span>, sharey=<span class="literal">False</span>)</span><br><span class="line">result = adata.uns[<span class="string">&#x27;rank_genes_groups&#x27;</span>]</span><br><span class="line">groups = result[<span class="string">&#x27;names&#x27;</span>].dtype.names</span><br><span class="line">pd.DataFrame(</span><br><span class="line">    &#123;group + <span class="string">&#x27;_&#x27;</span> + key[:<span class="number">1</span>] : result[key][group]</span><br><span class="line">    <span class="keyword">for</span> group <span class="keyword">in</span> groups <span class="keyword">for</span> key <span class="keyword">in</span> [<span class="string">&#x27;names&#x27;</span>, <span class="string">&#x27;pvals&#x27;</span>]&#125;</span><br><span class="line">).head(<span class="number">5</span>)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>marker基因展示</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 用小提琴图展示前20个差异基因在该簇和剩余簇的分布差异</span></span><br><span class="line"><span class="keyword">from</span> matplotlib.pyplot <span class="keyword">import</span> rc_context</span><br><span class="line"></span><br><span class="line"><span class="keyword">with</span> rc_context(&#123;<span class="string">&#x27;figure.figsize&#x27;</span>: (<span class="number">9</span>, <span class="number">1.5</span>)&#125;):</span><br><span class="line">    sc.pl.rank_genes_groups_violin(adata, n_genes=<span class="number">20</span>, jitter=<span class="literal">False</span>)</span><br><span class="line">    </span><br><span class="line"><span class="comment"># 直观地显示某个基因在不同簇中的表达</span></span><br><span class="line">sc.pl.violin(adata, [<span class="string">&#x27;NKG7&#x27;</span>], groupby=<span class="string">&#x27;leiden&#x27;</span>)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>细胞类型注释</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 每个cluster对应的细胞类型注释</span></span><br><span class="line">new_cluster_names = [</span><br><span class="line">    <span class="string">&#x27;0&#x27;</span>, <span class="string">&#x27;1&#x27;</span>, <span class="string">&#x27;2&#x27;</span>:<span class="string">&#x27;CD14&#x27;</span>, <span class="string">&#x27;3&#x27;</span>:<span class="string">&#x27;endo&#x27;</span>, <span class="string">&#x27;4&#x27;</span>, <span class="string">&#x27;5&#x27;</span>:<span class="string">&#x27;NK&#x27;</span></span><br><span class="line">]</span><br><span class="line"></span><br><span class="line">adata.rename_categories(<span class="string">&#x27;leiden&#x27;</span>, new_cluster_names)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 可视化结果展示</span></span><br><span class="line">pl.umap(adata, color=<span class="string">&#x27;leiden&#x27;</span>, legend_loc=<span class="string">&#x27;on data&#x27;</span>, title=<span class="string">&#x27;&#x27;</span>, frameon=<span class="literal">False</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 用tracksplot验证基因与细胞的关系</span></span><br><span class="line">sc.pl.rank_genes_groups_tracksplot(adata, n_genes=<span class="number">3</span>)</span><br></pre></td></tr></table></figure>
</li>
</ol>
<hr>
<h2 id="多个样本单细胞数据整合-ingest和BBKNN"><a class="headerlink" href="#多个样本单细胞数据整合-ingest和BBKNN"></a>多个样本单细胞数据整合 - ingest和BBKNN</h2>
<hr>
<h3 id="整合算法介绍"><a class="headerlink" href="#整合算法介绍"></a>整合算法介绍</h3>
<h4 id="ingest"><a class="headerlink" href="#ingest"></a>ingest</h4>
<ul>
<li>一种基于PCA的简单方法用于集成数据（integrating data）</li>
<li>用已知的细胞类型数据去注释未知的</li>
<li>ingest函数假设有一个带注释的参考数据集（<strong>有点类似SingleR</strong>），该数据集捕获了感兴趣的生物变异性。ingest是在参考数据上拟合模型，并使用它来投影新数据。目前，该模型是一个结合了邻居查找搜索树（neighbor lookup search tree）的PCA或者UMAP</li>
</ul>
<p>ingest有以下特点：</p>
<ol>
<li>由于ingest操作简单，过程清晰，因此工作流透明且快速</li>
<li>与BBKNN相似，ingest使数据矩阵本身保持不变。因为ingest是在embedding空间做分析</li>
<li>与BBKNN不同，ingest在解决了标签映射的同时，维持了可能具有特定簇或轨迹等所需属性的embedding（因为BBKNN是简单粗暴的对齐embedding）</li>
</ol>
<hr>
<h4 id="BBKNN"><a class="headerlink" href="#BBKNN"></a>BBKNN</h4>
<ul>
<li>BBKNN：Batch balanced KNN，BBKNN是一种快速直观的去除批次效应工具，可直接用于scanpy工作流。它是 <code>scanpy.pp.neighbors()</code> 的一种替代方法，这两个功能都是创建一个 neighbors graph，以便后续在聚类、pseudo time计算和UMAP可视化中使用</li>
<li>标准方法首先确定整个数据集中每个细胞的k个最近邻，然后将候选细胞转换为指数相关连接，然后作为进一步分析的基础。如果数据中存在批次效应（可能是由于不同的数据采集技术、协议变更，甚至是特别严重的操作员影响），那么将很难在不同批次中链接相应的细胞类型</li>
<li>BBKNN修改了邻域构建步骤，以生成一个跨所有批次数据平衡的图。这种方法将邻居网络视为数据的主要表示形式。对于每个细胞，BBKNN图是通过独立地为每个设备提供的批次（单个数据集）中的每个细胞找到k个最近邻居来构建的，这使得每个细胞在每个批次中都有一个独立的邻居池。同类细胞的这些邻居都被加入列表联系起来</li>
<li>注意：BBKNN会将列表中的<strong>元素用简单快速的方式对齐</strong>，即对齐不同批次的同类细胞。所以才说BBKNN实现了数据整合</li>
<li>BBKNN的主要假设是，<strong>批次间至少存在一些相同类型的细胞，并且批次效应引起的批次间相同细胞类型之间的差异小于批次内不同类型细胞之间的差异</strong>，在这种情况下，图的构造将跨批次将相似的细胞类型分组在一起，同时将不相关的细胞类型分开</li>
</ul>
<hr>
<h3 id="使用BBKNN进行数据整合"><a class="headerlink" href="#使用BBKNN进行数据整合"></a>使用BBKNN进行数据整合</h3>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> bbknn <span class="keyword">import</span> bbknn</span><br><span class="line"></span><br><span class="line"><span class="comment"># BBKNN整合</span></span><br><span class="line">sc.external.pp.bbknn(adatas, batch_key=<span class="string">&#x27;batch&#x27;</span>)</span><br><span class="line"></span><br><span class="line">sc.tl.umap(adatas)</span><br><span class="line">sc.pl.umap(adatas, color=[<span class="string">&#x27;batch&#x27;</span>, <span class="string">&#x27;orig&#x27;</span>])</span><br><span class="line"></span><br><span class="line">adatas.obsm</span><br><span class="line"></span><br><span class="line">sc.tl.leiden(adatas)</span><br></pre></td></tr></table></figure>
<hr>
<h2 id="基于scanorama整合scRNA-seq和空间数据分析"><a class="headerlink" href="#基于scanorama整合scRNA-seq和空间数据分析"></a>基于scanorama整合scRNA-seq和空间数据分析</h2>
<hr>
<p><strong>官方连接</strong>：<a target="_blank" rel="noopener" href="https://scanpy.readthedocs.io/en/stable/tutorials/spatial/">Spatial — scanpy</a></p>
<p><strong>与单细胞数据具体整合流程见</strong>：<a target="_blank" rel="noopener" href="https://scanpy.readthedocs.io/en/stable/tutorials/spatial/integration-scanorama.html">使用 scanorama 将空间数据与 scRNA-seq 集成 — scanpy</a></p>
<ul>
<li>Scanorama可以实现scRNA-seq数据集集成和批次矫正</li>
<li>文章提出的方法是整合scRNA-seq的高效方案，与过去的ingest不同，不需要指定某个具体的参考数据集，其实过去方法的思想还是将数据整合到一个更大的簇中（域适应到参考数据集）</li>
<li>Scanorama利用全景生成的思想，可以实现同时多个数据集的整合，可以自动匹配不同数据集下的相同细胞类型，同时保留不同数据集下不同细胞类型的差异，是更合理的整合方法</li>
<li>其中的异质指的是整合保留了两个数据集中真正存在差异的信息。Scanorama属于是实现了正确消除了技术，实验带来的批次效应</li>
</ul>
<p><strong>流程</strong>：</p>
<ol>
<li>
<p>导包</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> os</span><br><span class="line"><span class="keyword">import</span> anndata</span><br><span class="line"><span class="keyword">import</span> scanpy <span class="keyword">as</span> sc</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> matplotlib.pyplot <span class="keyword">as</span> plt</span><br><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"><span class="keyword">import</span> seaborn <span class="keyword">as</span> sns</span><br><span class="line"><span class="keyword">import</span> pooch</span><br><span class="line"><span class="keyword">from</span> scipy.sparse <span class="keyword">import</span> csr_matrix</span><br><span class="line"><span class="keyword">from</span> scipy.io <span class="keyword">import</span> mmwrite</span><br><span class="line"><span class="keyword">import</span> scanorama</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>设置一些全局参数</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">sc.logging.print_versions()</span><br><span class="line">sc.set_figure_params(facecolor=<span class="string">&quot;white&quot;</span>, figsize=(<span class="number">10</span>, <span class="number">10</span>))</span><br><span class="line">sc.settings.verbosity = <span class="number">3</span></span><br></pre></td></tr></table></figure>
</li>
<li>
<p>设置路径</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">print</span>(os.getcwd())</span><br><span class="line">os.chdir(<span class="string">&quot;/home/oelab/RunData2/STM_AD_analyze/data/GSE220442_counts_and_images/counts_and_images/&quot;</span>)</span><br><span class="line"><span class="built_in">print</span>(os.getcwd())</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>读取数据</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 读取数据</span></span><br><span class="line">AD1 = sc.read_visium(<span class="string">&quot;../counts_and_images/AD1/&quot;</span>)</span><br><span class="line">AD2 = sc.read_visium(<span class="string">&quot;../counts_and_images/AD2/&quot;</span>)</span><br><span class="line">AD3 = sc.read_visium(<span class="string">&quot;../counts_and_images/AD3/&quot;</span>)</span><br><span class="line">NC1 = sc.read_visium(<span class="string">&quot;../counts_and_images/NC1/&quot;</span>)</span><br><span class="line">NC2 = sc.read_visium(<span class="string">&quot;../counts_and_images/NC2/&quot;</span>)</span><br><span class="line">NC3 = sc.read_visium(<span class="string">&quot;../counts_and_images/NC3/&quot;</span>)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>质控计算</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 质控计算</span></span><br><span class="line">AD1.var_names_make_unique()</span><br><span class="line">AD2.var_names_make_unique()</span><br><span class="line">AD3.var_names_make_unique()</span><br><span class="line">NC1.var_names_make_unique()</span><br><span class="line">NC2.var_names_make_unique()</span><br><span class="line">NC3.var_names_make_unique()</span><br><span class="line"></span><br><span class="line">sc.pp.calculate_qc_metrics(AD1, inplace=<span class="literal">True</span>)</span><br><span class="line">sc.pp.calculate_qc_metrics(AD2, inplace=<span class="literal">True</span>)</span><br><span class="line">sc.pp.calculate_qc_metrics(AD3, inplace=<span class="literal">True</span>)</span><br><span class="line">sc.pp.calculate_qc_metrics(NC1, inplace=<span class="literal">True</span>)</span><br><span class="line">sc.pp.calculate_qc_metrics(NC2, inplace=<span class="literal">True</span>)</span><br><span class="line">sc.pp.calculate_qc_metrics(NC3, inplace=<span class="literal">True</span>)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>可视化</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 可视化</span></span><br><span class="line"><span class="keyword">for</span> name,adata <span class="keyword">in</span> [</span><br><span class="line">    (<span class="string">&quot;AD_sample1&quot;</span>, AD1),</span><br><span class="line">    (<span class="string">&quot;AD_sample2&quot;</span>, AD2),</span><br><span class="line">    (<span class="string">&quot;AD_sample3&quot;</span>, AD3),</span><br><span class="line">    (<span class="string">&quot;Normal_sample1&quot;</span>, NC1),</span><br><span class="line">    (<span class="string">&quot;Normal_sample2&quot;</span>, NC2),</span><br><span class="line">    (<span class="string">&quot;Normal_sample3&quot;</span>, NC3)]:</span><br><span class="line">    fig, axs = plt.subplots(<span class="number">1</span>, <span class="number">4</span>, figsize= (<span class="number">12</span>, <span class="number">3</span>))</span><br><span class="line">    fig.suptitle(<span class="string">f&quot;Covariates for filtering: <span class="subst">&#123;name&#125;</span>&quot;</span>)</span><br><span class="line"></span><br><span class="line">    sns.distplot(adata.obs[<span class="string">&quot;total_counts&quot;</span>], kde=<span class="literal">False</span>, ax = axs[<span class="number">0</span>])</span><br><span class="line">    </span><br><span class="line">    sns.distplot(</span><br><span class="line">        adata.obs[<span class="string">&quot;total_counts&quot;</span>][adata.obs[<span class="string">&quot;total_counts&quot;</span>] &lt; <span class="number">20000</span>],</span><br><span class="line">        kde=<span class="literal">False</span>, bins=<span class="number">40</span>, ax=axs[<span class="number">1</span>],)</span><br><span class="line">    </span><br><span class="line">    sns.distplot(</span><br><span class="line">        adata.obs[<span class="string">&quot;n_genes_by_counts&quot;</span>],</span><br><span class="line">        kde=<span class="literal">False</span>, bins=<span class="number">60</span>, ax=axs[<span class="number">2</span>],)</span><br><span class="line"></span><br><span class="line">    sns.distplot(</span><br><span class="line">        adata.obs[<span class="string">&quot;n_genes_by_counts&quot;</span>][adata.obs[<span class="string">&quot;n_genes_by_counts&quot;</span>] &lt; <span class="number">4000</span>],</span><br><span class="line">        kde=<span class="literal">False</span>, bins=<span class="number">60</span>, ax=axs[<span class="number">3</span>],)</span><br><span class="line">    plt.savefig(<span class="string">&quot;/home/oelab/RunData2/STM_AD_analyze/result/&quot;</span> + <span class="string">&quot;qc1.distplot_&quot;</span> + name + <span class="string">&quot;.png&quot;</span>)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>数据整合</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">AD1.obs[<span class="string">&quot;orig&quot;</span>] = <span class="string">&#x27;AD&#x27;</span></span><br><span class="line">AD2.obs[<span class="string">&quot;orig&quot;</span>] = <span class="string">&#x27;AD&#x27;</span></span><br><span class="line">AD3.obs[<span class="string">&quot;orig&quot;</span>] = <span class="string">&#x27;AD&#x27;</span></span><br><span class="line">NC1.obs[<span class="string">&quot;orig&quot;</span>] = <span class="string">&#x27;Normal&#x27;</span></span><br><span class="line">NC2.obs[<span class="string">&quot;orig&quot;</span>] = <span class="string">&#x27;Normal&#x27;</span></span><br><span class="line">NC3.obs[<span class="string">&quot;orig&quot;</span>] = <span class="string">&#x27;Normal&#x27;</span></span><br><span class="line"></span><br><span class="line">adatas = [AD1, AD2, AD3, NC1, NC2, NC3]</span><br><span class="line"></span><br><span class="line"><span class="comment"># 数据整合</span></span><br><span class="line">adatas_cor = scanorama.correct_scanpy(adatas, return_dimred=<span class="literal">True</span>)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>将多个数据整合到一起并保存在整合后的embeddings中</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">adatas_spatial = sc.concat(</span><br><span class="line">    adatas_cor, label = <span class="string">&quot;library_id&quot;</span>,</span><br><span class="line">    uns_merge=<span class="string">&quot;unique&quot;</span>, </span><br><span class="line">    keys=[</span><br><span class="line">        k</span><br><span class="line">        <span class="keyword">for</span> d <span class="keyword">in</span> [</span><br><span class="line">            adatas_cor[<span class="number">0</span>].uns[<span class="string">&quot;spatial&quot;</span>],</span><br><span class="line">            adatas_cor[<span class="number">1</span>].uns[<span class="string">&quot;spatial&quot;</span>],</span><br><span class="line">            adatas_cor[<span class="number">2</span>].uns[<span class="string">&quot;spatial&quot;</span>],</span><br><span class="line">            adatas_cor[<span class="number">3</span>].uns[<span class="string">&quot;spatial&quot;</span>],</span><br><span class="line">            adatas_cor[<span class="number">4</span>].uns[<span class="string">&quot;spatial&quot;</span>],</span><br><span class="line">            adatas_cor[<span class="number">5</span>].uns[<span class="string">&quot;spatial&quot;</span>]</span><br><span class="line">        ]</span><br><span class="line">        <span class="keyword">for</span> k, v <span class="keyword">in</span> d.items()</span><br><span class="line">    ],</span><br><span class="line">    index_unique=<span class="string">&quot;-&quot;</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>降维聚类</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">sc.pp.neighbors(adatas_spatial, use_rep=<span class="string">&quot;X_scanorama&quot;</span>)</span><br><span class="line">sc.tl.umap(adatas_spatial)</span><br><span class="line"></span><br><span class="line">sc.tl.leiden(adatas_spatial, key_added=<span class="string">&quot;clusters&quot;</span>)</span><br><span class="line"></span><br><span class="line">sc.pl.umap(</span><br><span class="line">    adatas_spatial, color=[<span class="string">&quot;clusters&quot;</span>, <span class="string">&quot;library_id&quot;</span>, <span class="string">&quot;orig&quot;</span>], palette=sc.pl.palettes.default_20</span><br><span class="line">)</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>空间可视化</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line">clusters_colors = <span class="built_in">dict</span>(</span><br><span class="line">    <span class="built_in">zip</span>([<span class="built_in">str</span>(i) <span class="keyword">for</span> i <span class="keyword">in</span> <span class="built_in">range</span>(<span class="number">18</span>)], adatas_spatial.uns[<span class="string">&quot;clusters_colors&quot;</span>])</span><br><span class="line">)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 开始画图</span></span><br><span class="line">fig, axs = plt.subplots(<span class="number">1</span>, <span class="number">6</span>, figsize= (<span class="number">15</span>, <span class="number">10</span>))</span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> i, library <span class="keyword">in</span> <span class="built_in">enumerate</span>(</span><br><span class="line">    [<span class="string">&quot;2-8&quot;</span>, <span class="string">&quot;T4857&quot;</span>, <span class="string">&quot;2-3&quot;</span>, <span class="string">&quot;2-5&quot;</span>, <span class="string">&quot;18-64&quot;</span>, <span class="string">&quot;1-1&quot;</span>]</span><br><span class="line">):</span><br><span class="line">    ad = adatas_spatial[adatas_spatial.obs.library_id == library,:].copy()</span><br><span class="line">    sc.pl.spatial(</span><br><span class="line">        ad, img_key=<span class="string">&quot;hires&quot;</span>, library_id=library,</span><br><span class="line">        color=<span class="string">&quot;clusters&quot;</span>, size=<span class="number">1.5</span>, palette=[</span><br><span class="line">            v</span><br><span class="line">            <span class="keyword">for</span> k,v <span class="keyword">in</span> clusters_colors.items()</span><br><span class="line">            <span class="keyword">if</span> k <span class="keyword">in</span> ad.obs.clusters.unique().tolist()],</span><br><span class="line">        legend_loc=<span class="literal">None</span>,</span><br><span class="line">        show=<span class="literal">False</span>,</span><br><span class="line">        ax = axs[i],</span><br><span class="line">    )</span><br><span class="line"></span><br><span class="line">plt.tight_layout()</span><br></pre></td></tr></table></figure>
</li>
</ol>
<hr>
<h2 id="Squidpy"><a class="headerlink" href="#Squidpy"></a>Squidpy</h2>
<hr>
<p><strong>教程</strong>：<a target="_blank" rel="noopener" href="https://squidpy.readthedocs.io/en/stable/">Squidpy - Spatial Single Cell Analysis in Python — squidpy documentation</a></p>
</article><div class="post-copyright"><div class="post-copyright__author_group"><a class="post-copyright__author_img" href="/about/"><img class="post-copyright__author_img_front" src="https://coo1heisenberg-blog.oss-cn-shanghai.aliyuncs.com/MyPhoto.jpg"></a><div class="post-copyright__author_name">coo1heisenberg's Blog</div><div class="post-copyright__author_desc">Diligence can make up for clumsiness!</div></div><div class="post-tools" id="post-tools"><div class="post-tools-left"><div id="quit-box" onclick="RemoveRewardMask()"></div></div></div><div class="post-copyright__notice"><span class="post-copyright-info">本文是原创文章，采用<a target="_blank" rel="noopener" href="https://creativecommons.org/licenses/by-nc-sa/4.0/deed.zh">CC BY-NC-SA 4.0</a>协议，完整转载请注明来自<a href="/">coo1heisenberg's Blog</a></span></div></div><div class="post-tools-right"><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/%E7%A9%BA%E9%97%B4%E8%BD%AC%E5%BD%95%E7%BB%84/"><span class="tags-punctuation"></span>空间转录组<span class="tagsPageCount">1</span></a></div></div><div class="social-share"></div></div><nav class="needEndHide pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2024/09/28/00-%E5%89%8D%E7%AB%AF%E5%BC%80%E5%8F%91%E5%85%A5%E9%97%A8%E6%95%99%E7%A8%8B/"><div class="pagination-info"><div class="label">上一篇</div><div class="prev_info">00 HTML5和CSS3 笔记</div></div></a></div><div class="next-post pull-right"><a href="/2024/09/13/00-%E5%8D%95%E7%BB%86%E8%83%9E%E5%A4%9A%E7%BB%84%E5%AD%A6%E5%88%86%E6%9E%90/"><div class="pagination-info"><div class="label">下一篇</div><div class="next_info">00 单细胞多组学分析 笔记</div></div></a></div></nav></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="card-content"><div class="card-info-avatar is-center"><div class="author-info__top-group"><div class="author-info__sayhi" id="author-info__sayhi" onclick="sco.changeSayHelloText()">sayhello.morning</div></div></div><div class="avatar-img-group"><img class="avatar-img" alt="头像" src="https://coo1heisenberg-blog.oss-cn-shanghai.aliyuncs.com/MyPhoto.jpg"><div class="avatar-sticker"><img class="avatar-sticker-img" src="https://7.isyangs.cn/34/65f2e4e0423cc-34.png" alt="心情贴纸"></div></div><div class="author-info__description_group"><div class="author-info__description">菜鸡的自我救赎...</div><div class="author-info__description2"></div></div><div class="author-info__bottom-group"><a class="author-info__bottom-group-left" href="/about/"><div class="author-info__name">Coo1heisenberg’s BLOG</div><div class="author-info__desc">Diligence can make up for clumsiness!</div></a><div class="card-info-social-icons is-center"><a class="social-icon" target="_blank" rel="noopener" href="https://github.com/coo1heisenbergProject" title="Github"><i class="solitude  st-github-line"></i></a></div></div></div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="solitude st-menu-line"></i><span>文章目录</span></div><div class="toc-content" id="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link"><span class="toc-text">空间转录组</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E7%A9%BA%E9%97%B4%E8%BD%AC%E5%BD%95%E7%BB%84%E7%9A%84%E6%B5%81%E7%A8%8B%E5%88%86%E6%9E%90"><span class="toc-text">空间转录组的流程分析</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Python%E5%9F%BA%E7%A1%80"><span class="toc-text">Python基础</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Python%E8%BF%9B%E9%98%B6"><span class="toc-text">Python进阶</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#DataFrame"><span class="toc-text">DataFrame</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#scverse%E5%88%86%E6%9E%90%E6%A1%86%E6%9E%B6-AnnData%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84"><span class="toc-text">scverse分析框架 - AnnData数据结构</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E7%A9%BA%E9%97%B4%E8%BD%AC%E5%BD%95%E7%BB%84%E5%9F%BA%E6%9C%AC%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B"><span class="toc-text">空间转录组基本分析流程</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#scanpy-%E5%8D%95%E4%B8%AA%E6%A0%B7%E6%9C%AC%E7%9A%84%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B"><span class="toc-text">scanpy 单个样本的单细胞分析流程</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%A4%9A%E4%B8%AA%E6%A0%B7%E6%9C%AC%E5%8D%95%E7%BB%86%E8%83%9E%E6%95%B0%E6%8D%AE%E6%95%B4%E5%90%88-ingest%E5%92%8CBBKNN"><span class="toc-text">多个样本单细胞数据整合 - ingest和BBKNN</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%95%B4%E5%90%88%E7%AE%97%E6%B3%95%E4%BB%8B%E7%BB%8D"><span class="toc-text">整合算法介绍</span></a><ol class="toc-child"><li class="toc-item toc-level-4"><a class="toc-link" href="#ingest"><span class="toc-text">ingest</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#BBKNN"><span class="toc-text">BBKNN</span></a></li></ol></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E4%BD%BF%E7%94%A8BBKNN%E8%BF%9B%E8%A1%8C%E6%95%B0%E6%8D%AE%E6%95%B4%E5%90%88"><span class="toc-text">使用BBKNN进行数据整合</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%9F%BA%E4%BA%8Escanorama%E6%95%B4%E5%90%88scRNA-seq%E5%92%8C%E7%A9%BA%E9%97%B4%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90"><span class="toc-text">基于scanorama整合scRNA-seq和空间数据分析</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Squidpy"><span class="toc-text">Squidpy</span></a></li></ol></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="solitude st-map-line"></i><span>最近发布</span></div><div class="aside-list"><div class="aside-list-item"><a class="thumbnail" href="/2024/11/04/Java-%E5%B9%B6%E5%8F%91%E7%BC%96%E7%A8%8B/" title="01 Java并发编程 笔记"><img alt="01 Java并发编程 笔记" src="https://coo1heisenberg-blog.oss-cn-shanghai.aliyuncs.com/Javalogobalala.png"></a><div class="content"><a class="title" href="/2024/11/04/Java-%E5%B9%B6%E5%8F%91%E7%BC%96%E7%A8%8B/" title="01 Java并发编程 笔记">01 Java并发编程 笔记</a><a class="article-recent_post_categories" href="/2024/11/04/Java-%E5%B9%B6%E5%8F%91%E7%BC%96%E7%A8%8B/">Java</a></div></div><div class="aside-list-item"><a class="thumbnail" href="/2024/11/01/Java-%E9%9B%86%E5%90%88%E6%A1%86%E6%9E%B6/" title="00 Java集合框架 笔记"><img alt="00 Java集合框架 笔记" src="https://coo1heisenberg-blog.oss-cn-shanghai.aliyuncs.com/Javalogobalala.png"></a><div class="content"><a class="title" href="/2024/11/01/Java-%E9%9B%86%E5%90%88%E6%A1%86%E6%9E%B6/" title="00 Java集合框架 笔记">00 Java集合框架 笔记</a><a class="article-recent_post_categories" href="/2024/11/01/Java-%E9%9B%86%E5%90%88%E6%A1%86%E6%9E%B6/">Java</a></div></div><div class="aside-list-item"><a class="thumbnail" href="/2024/09/28/00-%E5%89%8D%E7%AB%AF%E5%BC%80%E5%8F%91%E5%85%A5%E9%97%A8%E6%95%99%E7%A8%8B/" title="00 HTML5和CSS3 笔记"><img alt="00 HTML5和CSS3 笔记" src="https://coo1heisenberg-blog.oss-cn-shanghai.aliyuncs.com/Web_Html%2BCSS.png"></a><div class="content"><a class="title" href="/2024/09/28/00-%E5%89%8D%E7%AB%AF%E5%BC%80%E5%8F%91%E5%85%A5%E9%97%A8%E6%95%99%E7%A8%8B/" title="00 HTML5和CSS3 笔记">00 HTML5和CSS3 笔记</a><a class="article-recent_post_categories" href="/2024/09/28/00-%E5%89%8D%E7%AB%AF%E5%BC%80%E5%8F%91%E5%85%A5%E9%97%A8%E6%95%99%E7%A8%8B/">前端</a></div></div><div class="aside-list-item"><a class="thumbnail" href="/2024/09/26/01-%E7%A9%BA%E9%97%B4%E8%BD%AC%E5%BD%95%E7%BB%84/" title="01 空间转录组 笔记"><img alt="01 空间转录组 笔记" src="https://coo1heisenberg-blog.oss-cn-shanghai.aliyuncs.com/bio_gene.png"></a><div class="content"><a class="title" href="/2024/09/26/01-%E7%A9%BA%E9%97%B4%E8%BD%AC%E5%BD%95%E7%BB%84/" title="01 空间转录组 笔记">01 空间转录组 笔记</a><a class="article-recent_post_categories" href="/2024/09/26/01-%E7%A9%BA%E9%97%B4%E8%BD%AC%E5%BD%95%E7%BB%84/">生信分析</a></div></div><div class="aside-list-item"><a class="thumbnail" href="/2024/09/13/00-%E5%8D%95%E7%BB%86%E8%83%9E%E5%A4%9A%E7%BB%84%E5%AD%A6%E5%88%86%E6%9E%90/" title="00 单细胞多组学分析 笔记"><img alt="00 单细胞多组学分析 笔记" src="https://coo1heisenberg-blog.oss-cn-shanghai.aliyuncs.com/bio_gene.png"></a><div class="content"><a class="title" href="/2024/09/13/00-%E5%8D%95%E7%BB%86%E8%83%9E%E5%A4%9A%E7%BB%84%E5%AD%A6%E5%88%86%E6%9E%90/" title="00 单细胞多组学分析 笔记">00 单细胞多组学分析 笔记</a><a class="article-recent_post_categories" href="/2024/09/13/00-%E5%8D%95%E7%BB%86%E8%83%9E%E5%A4%9A%E7%BB%84%E5%AD%A6%E5%88%86%E6%9E%90/">生信分析</a></div></div></div></div></div></div></main><footer id="footer"><div id="st-footer-bar"><div class="footer-logo"><span>Solitude</span></div><div class="footer-bar-description">来自 Heisenberg 的原创文章</div><a class="footer-bar-link" href="/about/">了解更多</a></div><div id="footer_deal"></div><div id="st-footer"><div class="footer-group"><h3 class="footer-title">导航</h3><div class="footer-links"><a class="footer-item" href="/archives/" title="归档">归档</a><a class="footer-item" href="/categories/" title="分类">分类</a><a class="footer-item" href="/tags/" title="标签">标签</a></div></div><div class="footer-group"><h3 class="footer-title">服务</h3><div class="footer-links"><a class="footer-item" target="_blank" rel="noopener" href="https://aliyun.com/" title="阿里云">阿里云</a></div></div><div class="footer-group"><h3 class="footer-title">支持</h3><div class="footer-links"><a class="footer-item" href="/about/" title="打赏记录">打赏记录</a></div></div><div class="footer-group"><h3 class="footer-title">协议</h3><div class="footer-links"><a class="footer-item" href="/cookies/" title="Cookies">Cookies</a><a class="footer-item" href="/privacy/" title="用户协议">用户协议</a><a class="footer-item" href="/copyright/" title="版权协议">版权协议</a></div></div></div><div id="footer-bar"><div class="footer-bar-links"><div class="footer-bar-left"><div class="copyright">© 2024 By&nbsp;<a class="footer-bar-link" href="/">Coo1heisenberg’s BLOG</a></div></div><div class="footer-bar-right"><a class="footer-bar-link" target="_blank" rel="noopener" href="https://github.com/valor-x/hexo-theme-solitude" alt="主题">主题</a><a class="footer-bar-link cc" href="/null" aria-label="copyright"><i class="solitude st-copyright-line"></i><i class="solitude st-creative-commons-by-line"></i><i class="solitude st-creative-commons-nc-line"></i><i class="solitude st-creative-commons-nd-line"></i></a></div></div></div></footer></div><!-- right_menu--><!-- inject body--><div><script src="/js/utils.js?v=1.10.6"></script><script src="/js/main.js?v=1.10.6"></script><script src="/js/third_party/waterfall.min.js?v=1.10.6"></script><script src="//open.lightxi.com/cdnjs/ajax/libs/pjax/0.2.8/pjax.min.js"></script><script src="/js/third_party/universe.min.js?v=1.10.6"></script><script>dark()
</script><script src="//open.lightxi.com/cdnjs/ajax/libs/KaTeX/0.16.9/contrib/copy-tex.min.js"><script>(() => {
    document.querySelectorAll('#article-container span.katex-display').forEach(item => {
        utils.wrap(item, 'div', {class: 'katex-wrap'})
    })
})();
</script></script><script src="//open.lightxi.com/cdnjs/ajax/libs/node-snackbar/0.1.16/snackbar.min.js"></script><script src="//open.lightxi.com/cdnjs/ajax/libs/Swiper/11.0.5/swiper-bundle.min.js"></script><script>var meting_api = 'https://meting.qjqq.cn/?server=:server&type=:type&id=:id&auth=:auth&r=:r';</script><script src="//open.lightxi.com/cdnjs/ajax/libs/aplayer/1.10.1/APlayer.min.js"></script><script src="//open.lightxi.com/cdnjs/ajax/libs/meting/2.0.1/Meting.min.js"></script><script src="//open.lightxi.com/cdnjs/ajax/libs/pace/1.2.4/pace.min.js"></script><div class="js-pjax"><script defer pjax src="//open.lightxi.com/cdnjs/ajax/libs/busuanzi/2.3.0/bsz.pure.mini.min.js"></script></div></div><!-- newest comment--><!-- pjax--><script>const pjax = new Pjax({
    elements: 'a:not([target="_blank"])',
    selectors: ['title','#body-wrap','#site-config','meta[name="description"]','.js-pjax','meta[property^="og:"]','#config-diff'],
    cacheBust: false,
    analytics: false,
    scrollRestoration: false
})

document.querySelectorAll('script[data-pjax]').forEach(item => {
    const newScript = document.createElement('script')
    const content = item.text || item.textContent || item.innerHTML || ""
    Array.from(item.attributes).forEach(attr => newScript.setAttribute(attr.name, attr.value))
    newScript.appendChild(document.createTextNode(content))
    item.parentNode.replaceChild(newScript, item)
})

document.addEventListener('pjax:complete', () => {
    window.refreshFn()

    document.querySelectorAll('script[data-pjax]').forEach(item => {
        const newScript = document.createElement('script')
        const content = item.text || item.textContent || item.innerHTML || ""
        Array.from(item.attributes).forEach(attr => newScript.setAttribute(attr.name, attr.value))
        newScript.appendChild(document.createTextNode(content))
        item.parentNode.replaceChild(newScript, item)
    })

    GLOBAL_CONFIG.islazyload && window.lazyLoadInstance.update()

})

document.addEventListener('pjax:error', (e) => {
    if (e.request.status === 404) {
        pjax.loadUrl('/404.html')
    }
})</script><!-- theme--><script>initTheme = () => {
    let isDarkMode = window.matchMedia('(prefers-color-scheme: dark)').matches
    const cachedMode = utils.saveToLocal.get('theme');
    if (cachedMode === undefined) {
        const nowMode =
            isDarkMode ? 'dark' : 'light'
        document.documentElement.setAttribute('data-theme', nowMode);
    } else {
        document.documentElement.setAttribute('data-theme', cachedMode);
    }
    is_rm && rm.mode(cachedMode === 'dark' && isDarkMode)
}
initTheme()</script><!-- google adsense--><!-- search--><div id="local-search"><div class="search-dialog"><nav class="search-nav"><span class="search-dialog-title">搜索</span><button class="search-close-button"><i class="solitude st-close-fill"></i></button></nav><div class="search-wrap"><div class="search-box"><input class="search-box-input" id="search-input" type="text" autocomplete="off" spellcheck="false" autocorrect="off" autocapitalize="off" placeholder="输入关键词快速查找"></div><div id="search-results"><div id="search-hits"></div></div><div id="search-pagination"></div><div id="search-tips"></div></div></div><div id="search-mask"></div></div><script src="/js/search/local.js?v=1.10.6"></script><!-- Tianli-Talk--><!-- music--></body></html><script>const posts=["2024/11/04/Java-并发编程/","2024/11/01/Java-集合框架/","2024/09/28/00-前端开发入门教程/","2024/09/26/01-空间转录组/","2024/09/13/00-单细胞多组学分析/","2024/07/22/00-MongoDB/","2024/07/17/00-elasticsearch/","2024/07/15/01-黑马redis/","2024/07/13/00-黑马redis/","2024/07/08/00-黑马SSM/","2024/07/08/01-头条点评项目/","2024/07/06/00-Mybatis-Plus/","2024/06/29/00-黑马MySQL/","2024/06/25/00-头条点评项目/","2024/06/21/02-Python/","2024/06/11/01-Python/","2024/05/25/00-Python/","2024/05/13/DesignPattern/","2024/05/13/Docker/","2024/05/13/Nginx/","2024/05/13/Linux/"];function toRandomPost(){ pjax.loadUrl('/'+posts[Math.floor(Math.random()*posts.length)]); }</script>